---
title: "Regulatory Analysis"
author: "Andrew Burr"
date: "January 26, 2017"
output: html_document
---



```{r setup, include=FALSE}

require(limma)
require(knitr)
require(edgeR)
require(biomaRt)
require(DESeq2)
require(edgeR)
require(ggplot2)
require(reshape)
require(reshape2)
require(stringdist)
require(apcluster)
require(qqman)
require(Biostrings)
require(motifRG)
require(stringr)
require(scales)
require(NMF)


#force R to not use scientific notation
options(scipen=999)

#compute various information about the alignments.
alignmentInfo <- function(alignments){
  
  
  #find the distance alignment occured from 3 prime end of transcript
  #alignments$bp_end <- alignments$length - alignments$bp_start
  #count the number of mismatches
  #alignments$number_mismatches <- str_count(alignments$mismatches, ">")
  
  #finding total targets for each piRNA
  target_number <- as.data.frame(table(unique(subset(alignments, select = c(piRNA,hgnc_symbol)))[,1]))
  colnames(target_number) <- c("piRNA", "targets")
  alignments <- merge(alignments, target_number, by = "piRNA")
  
  #find total time gene is targeted by unique piRNA
  target_number <- as.data.frame(table(as.data.frame(unique(subset(alignments, select = c(piRNA,hgnc_symbol)))[,2])))
  colnames(target_number) <- c("hgnc_symbol", "gene_frequency")
  alignments <- merge(alignments, target_number, by = "hgnc_symbol")
  
  #denote weather the piRNA and/or the gene is differentially expressed
  alignments$DE_piRNA <- FALSE
  alignments$DE_piRNA[alignments$piRNA %in% rownames(voom_top)] <- TRUE

  alignments$DE_gene <- FALSE
  alignments$DE_gene[alignments$hgnc_symbol %in% top_genes$hgnc_symbol] <- TRUE

  #denote whether the 1T 10A bias is seen
  #alignments$`1T` <-  ifelse(sapply(alignments$piRNA, substr,1,1) == "T",TRUE,FALSE) 
  #alignments$`10A` <-  ifelse(sapply(alignments$piRNA, substr,10,10) == "A",TRUE,FALSE)
  #alignments$`1T10A` <- ifelse((alignments$`1T` & alignments$`10A`), TRUE, FALSE)
  
  
  target_number <- as.data.frame(table(unique(subset(alignments, select = c(piRNA,transcript_id)))[,2]))
  colnames(target_number) <- c("transcript", "targets")
  #rpkm <- merge(subset(alignments, select = c(transcript, length)) , target_number, by = "transcript")
  #rpkm$rpkm <- rpkm$targets/((rpkm$length)/1000)
  
  #alignments <- merge(alignments, subset(rpkm, select = c(transcript, rpkm)), by = "transcript")
  
  return(alignments)
}


removeRepeats <- function(read_counts){
  
  YRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt", header = FALSE, stringsAsFactors = FALSE)
  snoRNA <- read.table("../alignments/biogenesis_alignments_norRNA_25_33/small_sa_og_norRNA_alignments_snoRNA.txt", header = FALSE, stringsAsFactors = FALSE)
  
  rmsk_plus40 <- read.table("../alignments/small_sa_og_norRNA_alignments_repeatmasker_plus40.txt", header = FALSE, stringsAsFactors = FALSE)
  rmsk_plus40_tRNA <- rmsk_plus40[grepl(pattern = "tRNA", rmsk_plus40$V2),]
  rmsk <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker.txt", header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  
  rmsk$Family <- ifelse(grepl("tRNA", rmsk$V2), "tRNA", 
                        ifelse(grepl(")n", rmsk$V2), "Simple", 
                               ifelse(grepl("Alu", rmsk$V2), "Alu", 
                                      ifelse(rmsk$V2 %in% c("U1", "U2", "U3", "U4", "U5", "U6", "U7", "U8"), "U", 
                                             ifelse(grepl("HY", rmsk$V2), "Y-RNA", 
                                                    ifelse(grepl(pattern =  paste(c("L1", "L2", "L3", "L4", "L5", "L6"), collapse = "|"), rmsk$V2), "LINE", 
                                                           ifelse(grepl("MIR", rmsk$V2), "MIR", "other" )))))))
  
  rmsk$Family[rmsk$V2 == "5S"] <- "5S"
  rmsk$Family[rmsk$V2 == "7SLRNA"] <- "7SLRNA"
  rmsk$Family[rmsk$V2 == "7SK"] <- "7SK"
  
  
  
  
  
  
  #read_counts <- read_counts[!(rownames(read_counts) %in% tRNA$V1),]
  #read_counts <- read_counts[!(rownames(read_counts) %in% lncRNA$V1),]
  read_counts <- read_counts[!(rownames(read_counts) %in% YRNA$V1),]
  #read_counts <- read_counts[!(rownames(read_counts) %in% snoRNA$V1),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "Simple"]),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "Alu"]),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "5S"]),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "7SLRNA"]),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "7SK"]),]
  read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1[rmsk$Family == "other"]),]
  
  return(read_counts)
}



```



*Alignment Analysis*

Below are the code chunks for analyzing the alignments gathered between each piRNA and the targeting sequences.



```{r coexpression_setup}



#loads the targeting sequences expressionts
voom_sRNA <- read.table(file = "../data/targeting_sRNA.csv",
                        stringsAsFactors = F, 
                        header  = T, 
                        sep = ",")
rmsk <- read.table(file = "../alignments/small_sa_og_norRNA_25_33_25bp_alignments_repeatmasker.txt",
                   stringsAsFactors = F,
                   header  = F,
                   sep = "\t")
rmsk <- rmsk[!grepl(rmsk$V3, pattern = "tRNA"),]
voom_sRNA <- voom_sRNA[!(voom_sRNA$sRNA %in% rmsk$V1),]
voom_top <- read.table(file = "../data/top_targeting_sRNA.csv", 
                       stringsAsFactors = F,
                       header  = T, 
                       sep = ",")



conversion <- read.table("../../conversion_sets/ensembl_conversion.txt", header = T, stringsAsFactors = F, sep = "\t")

#load in the gene expression from RNA seq experiment
gene_expression <- read.csv(file = "../../RNAseq/data/GSC_vs_nonGSC_all_differentially_expressed_genes.csv", 
                            sep = ",", 
                            stringsAsFactors = F)
top_genes <- read.csv(file = "../../RNAseq/data/GSC_vs_nonGSC_top_pval05_differentially_expressed_genes.csv", 
                      sep = ",",
                      stringsAsFactors = F)


cDNA_alignments <- list()
cDNA_alignments[[1]] <-  read.delim("../data/cDNA_temp_alignments_bt2v1score2.txt", header=FALSE, stringsAsFactors=FALSE)[,]
cDNA_alignments[[2]] <-  read.delim("../data/cDNA_temp_alignments_bt2v1score4.txt", header=FALSE, stringsAsFactors=FALSE)[,]
cDNA_alignments[[3]] <-  read.delim("../data/cDNA_temp_alignments_bt2v1score6.txt", header=FALSE, stringsAsFactors=FALSE)[,]
#cDNA_alignments[[4]] <-  read.delim("../data/cDNA_temp_alignments_bt2v1score8.txt", header=FALSE, stringsAsFactors=FALSE)[,]
#cDNA_alignments[[5]] <-  read.delim("../data/cDNA_temp_alignments_bt2v1score10.txt", header=FALSE, stringsAsFactors=FALSE)[,]



UTR_alignments <- list()
UTR_alignments[[1]] <-  read.delim("../data/3UTR_temp_alignments_bt2v1score2.txt", 
                                   header=FALSE, 
                                   stringsAsFactors=FALSE)[,]
UTR_alignments[[2]] <-  read.delim("../data/3UTR_temp_alignments_bt2v1score4.txt", header=FALSE, stringsAsFactors=FALSE)[,]
UTR_alignments[[3]] <-  read.delim("../data/3UTR_temp_alignments_bt2v1score6.txt", header=FALSE, stringsAsFactors=FALSE)[,]
UTR_alignments_names <- c("bt2v1score2", "bt2v1score4" , "bt2v1score6")


# WHERE ARE THESE?!?!?!!??!?!?!
# 
# gsgarahd
# sfhsd
# sghfg
# sgsh
# sghshhsrhjsytu
# sgsgs
# liujkh;lkugs

promoter_alignments <- list()
promoter_alignments[[1]] <-  read.delim("../data/promoter_temp_alignments_bt2v1score2.txt", header=FALSE, stringsAsFactors=FALSE)[,]
promoter_alignments[[2]] <-  read.delim("../data/promoter_temp_alignments_bt2v1score4.txt", header=FALSE, stringsAsFactors=FALSE)[,]
promoter_alignments[[3]] <-  read.delim("../data/promoter_temp_alignments_bt2v1score6.txt", header=FALSE, stringsAsFactors=FALSE)[,]
promoter_alignments_names <- c("bt2v1score2", "bt2v1score4" , "bt2v1score6")


```



**3UTR**


```{r 3UTR_alignments, echo = FALSE}


#loading in the bowtie alignments from piRNA against cDNA
#alignments <- read.delim("../alignments/targeting_alignments/cDNA/sa_og_norRNA_25_33_noncRNA_21bp_list_alignments_3UTR_v3.txt", header=FALSE, stringsAsFactors=FALSE)[,]

#system("grep -f temp_read_names.txt ../alignments/list_targeting/small_sa_og_norRNA_25_33_25bp_targeting_3UTR_51_l9_n0_e140.txt > 3UTR_temp_alignments.txt" )

#alignments <- read.delim("3UTR_temp_alignments.txt", header=FALSE, stringsAsFactors=FALSE)[,]
#label the columns approriately
#colnames(alignments) <- c("piRNA", "gene", "bp_start", "mismatches")


for(alignments_number in 1:length(UTR_alignments)){
  
  alignments <- as.data.frame(UTR_alignments[[alignments_number]])
  name <- UTR_alignments_names[alignments_number]
  
  colnames(alignments) <- c("piRNA", "transcript_id", "bp_start", "alignment_info", "mismatches")
  #alignments$transcript_id <- gsub(pattern = "\\..*", replacement = "", alignments$transcript_id)
  alignments$transcript_id <- sapply(alignments$transcript_id,
                                     FUN = function(x){unlist(strsplit(x, split = "\\|"))[2]})
  
  conversion <- read.table(file =  "../../conversion_sets/ensembl_conversion.txt",
                           header = T, 
                           stringsAsFactors = F,
                           sep = "\t")

  alignments <- merge(alignments, 
                      conversion, 
                      by = "transcript_id" )
  alignments <- unique(alignments)
  alignments <- alignmentInfo(alignments)
  
  #alignments <- alignments[alignments$targets < 20,]
  
  
  
  coexpression <- 
    unique(
      merge(subset(voom_sRNA, 
                   select = c(sRNA, logFC, adj.P.Val)),
            subset(alignments, 
                   select = c(piRNA, hgnc_symbol)),
            by.x = "sRNA", by.y = "piRNA"))
  
  colnames(coexpression)[1:3] <- c("piRNA", "piRNA_logFC", "piRNA_pval")
  coexpression <-merge(coexpression,
                       subset(gene_expression, 
                              select = c(hgnc_symbol, logFC, adj.P.Val)),
                       by = "hgnc_symbol")
  colnames(coexpression)[5:6] <- c( "gene_logFC", "gene_pval")
  
  
  
  write.table(x = coexpression,
              file =  paste("../data/3UTR_coexpression_", name ,".csv"),
              quote = F,
              sep = ",",
              row.names = F)
  
  
  # create and save the coexpression plot
  pdf(file =  paste("../../figures/3UTR_coexpression_",name ,".pdf"), 6,4)
  print(
    
    #create the correlation plot 
    ggplot(data = coexpression, aes(y = gene_logFC, 
                                    x = piRNA_logFC)) +
      # geom_rect(aes(xmin = 2, xmax =ceiling(max(coexpression$piRNA_logFC)), ymax = -2, ymin = floor(min(coexpression$gene_logFC))), fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      # geom_rect(aes(xmax = -2, xmin = floor(min(coexpression$piRNA_logFC)), ymin = 2, ymax = ceiling(max(coexpression$gene_logFC))),fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      geom_point(aes(color = gene_pval, 
                     y = gene_logFC,
                     x = piRNA_logFC, 
                     alpha = 1-piRNA_pval), 
                 size = .9) +
      scale_color_gradientn(name ="Gene Pval" , 
                            colors = c("darkblue","lightblue","red"),
                            breaks = c( .01,.1,1), 
                            limits = c(0,1), 
                            values = c(.0000001, .1, 1)) +
      scale_alpha_continuous(labels = c(1,.1,.05,.001), 
                             limits = c(.001,  1),
                             breaks = c( .001,.05,.1,1),
                             name = "piRNA Pval") + 
      geom_hline(aes(yintercept = 2), 
                 linetype ="dotted") +
      geom_hline(aes(yintercept = -2), 
                 linetype = "dotted") +
      geom_vline(aes(xintercept = 2), 
                 linetype ="dotted") +
      geom_vline(aes(xintercept = -2), 
                 linetype ="dotted") +
      scale_y_continuous(breaks = c(seq((floor(min(coexpression$gene_logFC))), ceiling(max(coexpression$gene_logFC)), 1))) +
      scale_x_continuous(breaks = c(seq((floor(min(coexpression$piRNA_logFC))), ceiling(max(coexpression$piRNA_logFC)), 1))) +
      #scale_linetype_manual(guide = FALSE) +
      
      guides(linetype = FALSE) +   
      #ggtitle("3UTR Targeting: Gene LFC ~ piRNA LFC ") +
      #scale_color_manual(values = c("black", "blue")) + 
      xlab("piRNA logFC") +
      ylab("Target logFC") +
      #scale_size(range =c(.5,1.5)) + 
      # labs(  color = "Gene Pval") +
      theme_bw() 
      #theme(panel.grid.major.x = element_blank()) 
      #theme(legend.position ="right")
    
    
  )
  dev.off()
}

```




**cDNA**


```{r cDNA_alignments, echo = FALSE}


# file to convert ensembl gene ID to gene name
conversion <- read.table("../../conversion_sets/ensembl_conversion.txt", header = T, stringsAsFactors = F, sep = "\t")
#top_genes <- read.csv("../../RNAseq/data/StemVsDiff_top_DE_genes.txt", stringsAsFactors=FALSE)[,c(1,2,8)]
alignments_names <- c("gene_bt2v1score2", "gene_bt2v1score4" , "gene_bt2v1score6", "gene_bt2v1score8", "gene_bt2v1score10")
gene_expression <- read.csv("../../RNAseq/data/GSC_vs_nonGSC_all_differentially_expressed_genes.csv", sep = ",", stringsAsFactors = F)

gene_expression <- 
  unique(
    merge(conversion[,c(1,3)],
          gene_expression, 
          by.x = "gene_id",
          by.y = "row.names")
  )

for(alignments_number in 1:length(cDNA_alignments)){
  # load in the alignments
  alignments <- as.data.frame(cDNA_alignments[[alignments_number]])
  name <- alignments_names[alignments_number]
  
  colnames(alignments) <- c("piRNA", "transcript_id", "bp_start", "alignment_info", "mismatches")
  # remove the ".#" descriptor (not needed) from transcript ID
  alignments$transcript_id <- gsub(pattern = "\\..*",
                                   replacement = "", 
                                   alignments$transcript_id)
  # convert the ID to the gene name
  alignments <- merge(alignments, 
                      conversion, 
                      by = "transcript_id" )
  # select only unique gene alignments
  alignments <- unique(alignments)
  
  #split the alignment reference name into the appropraite columns
  #alignments$hgnc_symbol <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[4]})
  #alignments$transcript <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[1]})
  #alignments$length <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[5]})
  #alignments$length <-  as.numeric(sapply(alignments$length, sub, pattern =  "transcript_length:", replacement =  ""))
  
  #generate alignment info
  alignments <- alignmentInfo(alignments)
  
  
  #alignments <- alignments[alignments$targets < 20,]
  
  
  #create the coexpression table
  coexpression <-
    unique(
      merge(
        subset(x = voom_sRNA,
               select = c(sRNA, logFC, adj.P.Val)),
        subset(x = alignments, 
               select = c(piRNA, hgnc_symbol)),
        by.x = "sRNA",
        by.y = "piRNA"))
  
  colnames(coexpression)[1:3] <- c("piRNA", "piRNA_logFC", "piRNA_pval")
  coexpression <-merge(coexpression,
                       subset(gene_expression, 
                              select = c(hgnc_symbol, logFC, adj.P.Val)),
                       by = "hgnc_symbol")
  colnames(coexpression)[5:6] <- c( "gene_logFC", "gene_pval")
  # save the coexpressions to data directory
  write.table(x = coexpression, 
              file =  paste("../data/cDNA_coexpression_", name ,".csv"),
              quote = F,
              sep = ",", 
              row.names = F)
  # write the plot to figures directory
  pdf(paste("../../figures/cDNA_coexpression_",name ,".pdf"), 6,4)
  print(
    
    ggplot(data = coexpression,aes( y = gene_logFC, 
                                    x = piRNA_logFC)) +
      #geom_rect(aes(xmin = 2, xmax =ceiling(max(coexpression$piRNA_logFC)), ymax = -2, ymin = floor(min(coexpression$gene_logFC))), fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      #geom_rect(aes(xmax = -2, xmin = floor(min(coexpression$piRNA_logFC)), ymin = 2, ymax = ceiling(max(coexpression$gene_logFC))),fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      geom_point(aes(color = gene_pval, 
                     y = gene_logFC,
                     x = piRNA_logFC, 
                     alpha = 1-piRNA_pval), 
                 size = .9) +
      scale_color_gradientn(name ="Gene Pval" , 
                            colors = c("darkblue","lightblue","red"),
                            breaks = c( .01,.1,1),
                            limits = c(0,1), 
                            values = c(.0000001, .1, 1)) +
      scale_alpha_continuous(labels = c(1,.1,.05,.001), 
                             limits = c(.001,  1), 
                             breaks = c( .001,.05,.1,1), 
                             name = "piRNA Pval") + 
      geom_hline(aes(yintercept = 2), 
                 linetype ="dotted") +
      geom_hline(aes(yintercept = -2), 
                 linetype = "dotted") +
      geom_vline(aes(xintercept = 2), 
                 linetype ="dotted") +
      geom_vline(aes(xintercept = -2), 
                 linetype ="dotted") +
      scale_y_continuous(breaks = c(seq((floor(min(coexpression$gene_logFC))), ceiling(max(coexpression$gene_logFC)), 1))) +
      scale_x_continuous(breaks = c(seq((floor(min(coexpression$piRNA_logFC))), ceiling(max(coexpression$piRNA_logFC)), 1))) +
      #scale_linetype_manual(guide = FALSE) +
      
      guides(linetype = FALSE) +   
      #ggtitle("mRNA Targeting: Gene LFC ~ piRNA LFC ") +
      #scale_color_manual(values = c("black", "blue")) + 
      xlab("piRNA logFC") +
      ylab("Target logFC") +
      #scale_size(range =c(.5,1.5)) + 
      # labs(  color = "Gene Pval") +
      theme_bw() 
      #theme(panel.grid.major.x = element_blank()) +
      #theme(legend.position ="none")
    

  )
  dev.off()
}

gene_expression <- read.csv(file =  "../../RNAseq/data/GSC_vs_nonGSC_all_differentially_expressed_transcripts.csv",
                            stringsAsFactors = F)
top_genes <- read.csv(file = "../../RNAseq/data/GSC_vs_nonGSC_top_pval<05_differentially_expressed_transcripts.csv", 
                      stringsAsFactors=FALSE)[,c(1,2,8)]
alignments_names <- c("transcript_bt2v1score2", "transcript_bt2v1score4" , "transcript_bt2v1score6")




for(alignments_number in 1:length(cDNA_alignments)){
  
  alignments <- as.data.frame(cDNA_alignments[[alignments_number]])
  name <- alignments_names[alignments_number]
  
  colnames(alignments) <- c("piRNA", "transcript_id", "bp_start", "alignment_info", "mismatches")
  alignments$transcript_id <- gsub(pattern = "\\..*", 
                                   replacement = "", 
                                   alignments$transcript_id)
  
  alignments <- merge(alignments, 
                      conversion, 
                      by = "transcript_id" )
  
  alignments <- unique(alignments)
  
  #split the alignment reference name into the appropraite columns
  #alignments$hgnc_symbol <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[4]})
  #alignments$transcript <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[1]})
  #alignments$length <- sapply(alignments$gene, FUN = function(x){unlist(strsplit(x, split = " "))[5]})
  #alignments$length <-  as.numeric(sapply(alignments$length, sub, pattern =  "transcript_length:", replacement =  ""))
  
  #generate alignment info
  alignments <- alignmentInfo(alignments)
  
  
  #alignments <- alignments[alignments$targets < 20,]
  
  
  #create the coexpression table
  coexpression <- unique(merge(subset(voom_sRNA, select = c(sRNA, logFC, adj.P.Val)),
                               subset(alignments, select = c(piRNA, hgnc_symbol,transcript_id)),
                               by.x = "sRNA", by.y = "piRNA"))
  colnames(coexpression)[1:3] <- c("piRNA", "piRNA_logFC", "piRNA_pval")
  coexpression <-merge(coexpression,
                       subset(gene_expression, select = c(transcript_id, logFC, adj.P.Val)),
                       by = "transcript_id")
  colnames(coexpression)[6:7] <- c( "gene_logFC", "gene_pval")
  
 write.table(coexpression, paste("../data/cDNA_coexpression_", name ,".csv"), quote = F, sep = ",",row.names = F)
  
  pdf(paste("../../figures/cDNA_coexpression_",name ,".pdf"), 6,4)
  print(
    
    ggplot(data = coexpression,aes(y = gene_logFC, 
                                   x = piRNA_logFC)) +
      #geom_rect(aes(xmin = 2, xmax =ceiling(max(coexpression$piRNA_logFC)), ymax = -2, ymin = floor(min(coexpression$gene_logFC))), fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      #geom_rect(aes(xmax = -2, xmin = floor(min(coexpression$piRNA_logFC)), ymin = 2, ymax = ceiling(max(coexpression$gene_logFC))),fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      geom_point(aes(color = gene_pval, 
                     y = gene_logFC, 
                     x = piRNA_logFC, 
                     alpha = 1-piRNA_pval), 
                 size = .9) +
      scale_color_gradientn(name ="Gene Pval" ,
                            colors = c("darkblue", "lightblue", "red"),
                            breaks = c(.01, .1, 1), 
                            limits = c(0, 1),
                            values = c(.0000001, .1, 1)) +
      scale_alpha_continuous( labels = c(1,.1,.05,.001),
                              limits = c(.001, 1),
                              breaks = c(.001, .05, .1, 1), 
                              name = "piRNA Pval") + 
      geom_hline(aes(yintercept = 2), 
                 linetype ="dotted") +
      geom_hline(aes(yintercept = -2), 
                 linetype = "dotted") +
      geom_vline(aes(xintercept = 2), 
                 linetype ="dotted") +
      geom_vline(aes(xintercept = -2), 
                 linetype ="dotted") +
      scale_y_continuous(breaks = c(seq((floor(min(coexpression$gene_logFC))), ceiling(max(coexpression$gene_logFC)), 1))) +
      scale_x_continuous(breaks = c(seq((floor(min(coexpression$piRNA_logFC))), ceiling(max(coexpression$piRNA_logFC)), 1))) +
      #scale_linetype_manual(guide = FALSE) +
      
      guides(linetype = FALSE) +  
      #ggtitle("mRNA Targeting: Isoform LFC ~ piRNA LFC ") +
      #scale_color_manual(values = c("black", "blue")) + 
      xlab("piRNA logFC") +
      ylab("Target logFC") +
      #scale_size(range =c(.5,1.5)) + 
      # labs(  color = "Gene Pval") +
      theme_bw() 
      #theme(panel.grid.major.x = element_blank()) +
      #theme(legend.position ="none")
    

  )
  dev.off()
  
  
}


```





**Promoter**

```{r promotor_alignments, echo = FALSE}

#conversion table for converting RefSeq ids to HGNC gene symbols
conversion <- read.delim("../../conversion_sets/hgnc_to_refseq.txt", stringsAsFactors=FALSE, sep = "\t")
colnames(conversion) <- c("hgnc_symbol", "refseq_id")



for(alignments_number in 1:length(promoter_alignments)){
  
  alignments <- as.data.frame(promoter_alignments[[alignments_number]])
  name <- promoter_alignments_names[alignments_number]
  
  colnames(alignments) <- c("piRNA", "refseq_id", "bp_start", "alignment_info", "mismatches")
  #alignments$transcript_id <- gsub(pattern = "\\..*", replacement = "", alignments$transcript_id)
  
  alignments <- merge(alignments, 
                      conversion, 
                      by = "refseq_id")
  alignments <- unique(alignments)
  
  #finding total targets for each piRNA
  target_number <- 
    as.data.frame(
      table(
        unique(
          subset(x = alignments, 
                 select = c(piRNA,hgnc_symbol)
          )
        )[,1]
      )
    )
  colnames(target_number) <- c("piRNA", "targets")
  alignments <- merge(alignments, target_number, by = "piRNA")
  
  #find total time gene is targeted by unique piRNA
  target_number <-
    as.data.frame(
      table(
        unique(
          subset(x = alignments, 
                 select = c(piRNA,hgnc_symbol)
          )
        )[,2]
      )
    )
  
  
  colnames(target_number) <- c("hgnc_symbol", "gene_frequency")
  alignments <- merge(alignments,
                      target_number,
                      by = "hgnc_symbol")
  
  
  #create the coexpression table
  coexpression <- 
    unique(
      merge(
        subset(x = voom_sRNA, 
               select = c(sRNA, logFC, adj.P.Val)
        ),
        subset(x = alignments,
               select = c(piRNA, hgnc_symbol)
        ),
        by.x = "sRNA",
        by.y = "piRNA")
    )
  
  
  colnames(coexpression)[1:3] <- c("piRNA", "piRNA_logFC", "piRNA_pval")
  coexpression <- merge(coexpression,
                        subset(x = gene_expression, 
                               select = c(hgnc_symbol, logFC, adj.P.Val)
                        ),
                        by = "hgnc_symbol")
  colnames(coexpression)[5:6] <- c( "gene_logFC", "gene_pval")
  
  write.table(x = coexpression,
              paste("../data/promoter_coexpression_", name ,".csv"),
              quote = F, 
              sep = ",",
              row.names = F)
  
  pdf(paste("../../figures/promoter_coexpression_",name ,".pdf"), 6,4)
  print(
    #plot of the correlatiosn between expression data
    ggplot(data = coexpression, aes(y = gene_logFC, 
                                    x = piRNA_logFC)) +
      # geom_rect(aes(xmin = 2, xmax =ceiling(max(coexpression$piRNA_logFC)), ymax = -2, ymin = floor(min(coexpression$gene_logFC))), fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      #geom_rect(aes(xmax = -2, xmin = floor(min(coexpression$piRNA_logFC)), ymin = 2, ymax = ceiling(max(coexpression$gene_logFC))),fill = "white", color = "blue", alpha = 1/20000, size = .7) + 
      geom_point(aes(color = gene_pval, 
                     y = gene_logFC, 
                     x = piRNA_logFC, 
                     alpha = 1-piRNA_pval), 
                 size = .9) +
      scale_color_gradientn(name ="Gene Pval" , 
                            colors = c("darkblue","lightblue","red"),
                            breaks = c( .01,.1,1), 
                            limits = c(0,1), 
                            values = c(.0000001, .1, 1)) +
      scale_alpha_continuous(labels = c(1,.1,.05,.001),
                             limits = c(.001,  1), 
                             breaks = c( .001,.05,.1,1), 
                             name = "piRNA Pval") + 
      geom_hline(aes(yintercept = 2), 
                 linetype ="dotted") +
      geom_hline(aes(yintercept = -2),
                 linetype = "dotted") +
      geom_vline(aes(xintercept = 2),
                 linetype ="dotted") +
      geom_vline(aes(xintercept = -2),
                 linetype ="dotted") +
      scale_y_continuous(breaks = c(seq((floor(min(coexpression$gene_logFC))), ceiling(max(coexpression$gene_logFC)), 1))) +
      scale_x_continuous(breaks = c(seq((floor(min(coexpression$piRNA_logFC))), ceiling(max(coexpression$piRNA_logFC)), 1))) +
      #scale_linetype_manual(guide = FALSE) +
      
      guides(linetype = FALSE) +   
      #ggtitle("Promoter Region Targeting: Gene LFC ~ piRNA LFC ") +
      #scale_color_manual(values = c("black", "blue")) + 
      xlab("piRNA LFC") +
      ylab("Target LFC") +
      #scale_size(range =c(.5,1.5)) + 
      # labs(  color = "Gene Pval") +
      theme_bw() 
    #theme(panel.grid.major.x = element_blank()) +
    #   theme(legend.position ="none")
  )
  dev.off()
}



```
