---
title: "sRNA_differential_expression"
author: "Andrew Burr"
date: "April 23, 2017"
output: html_document
---



```{r setup, include=FALSE}
source('global.R')
#force R to not use scientific notation
options(scipen=999)


data_directory = '~/MEGA/data_files'

alignments_directory = file.path(data_directory, 'stem_diff_alignments')

#load in the gene expression from RNA seq experiment
gene_expression <- read.csv(file = "../../RNAseq/data/StemVsDiff_all_DE_genes.txt", 
                            stringsAsFactors = F)
transcript_expression <- read.csv(file = "../../RNAseq/data/StemVsDiff_all_DE_transcripts.csv", 
                                  stringsAsFactors = F)
#load in the differentially expressed genes info from the RNA seq experiment
top_genes <- read.csv(file = "../../RNAseq/data/StemVsDiff_top_DE_genes.txt",
                      stringsAsFactors=FALSE)

conversion <- read.table("../../conversion_sets//ensembl_conversion.txt", header = T, stringsAsFactors = F, sep = "\t")



```


**piRNA Differential Expression**

```{r piRNA_differential_expression, echo = FALSE}

# TODO replace with dplyr stuff
# TODO create expressions with no tRNA
#import the desired read counts
read.counts <- read.csv(file = file.path(data_directory, "sRNA_counts/small_sa_og_norRNA_25_33_counts.txt"), 
                        stringsAsFactors=FALSE,
                        row.names = 1)[,]

#remove non used reads
rRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_rRNA.txt"),
                   header = FALSE, 
                   stringsAsFactors = FALSE)

read.counts <- read.counts[!(rownames(read.counts) %in% rRNA$V1),]
YRNA <- read.table(file = file.path(allignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt"),
                   header = FALSE, 
                   stringsAsFactors = FALSE)
read.counts <- read.counts[!(rownames(read.counts) %in% YRNA$V1),]

#keep only cell lines from experiment
read.counts <- read.counts[,4:19]
#switch the cell lines
read.counts <- read.counts[,c(9:16, 1:8)]

read.counts <- read.counts[,-c(3,11)]

read.counts <- read.counts[!rowSums(read.counts) == 0,]

#read.counts <- removeRepeats(read.counts)
#remove low counts for differential expression
read.counts <- read.counts[rowSums(cpm(read.counts) >= 1) >= 7,]
read.counts <- read.counts[!rowSums(read.counts) == 0,]




#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge.counts <- 
  read.counts  %>% 
  DGEList() %>%
  calcNormFactors(method = c('TMM'))
#dge.counts <- dge.counts[rowSums(cpm(dge.counts) >=1) >=7,]

#voom normalization
voom.counts <- cpm(x = dge.counts,
                   log = TRUE, 
                   prior.count = 3)
#MDSplot
plotMDS(dge.counts, labels = colnames(read.counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the counts
fit <-
  voom.counts %>% 
  lmFit(design = design) %>%
  eBayes(trend = TRUE)


#voom fold changesmiRNA_alignments
voom.sRNA <- topTable(fit = fit, 
                      coef = "celltypeS", 
                      sort.by ="logFC", 
                      number = nrow(voom.counts) )
voom.top <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by ="logFC", 
                     number = nrow(voom.counts),
                     lfc = 0,
                     p.value = 0.05)
#voomNormExp <- voomCounts$E
#rm(celltype);rm(design.array); rm(design); rm(dge.counts)
#my name function

#adding the reads names into its own column for merging
voom.sRNA$sRNA <- rownames(voom.sRNA)
voom.top.sRNA <- voom.top
voom.top.sRNA$sRNA <- rownames(voom.top.sRNA)

#voom_top_sRNA$sRNA <- strtrim(voom_top_sRNA$sRNA, 25)
#write.table("../fastx_files/DE_piRNA_list.txt",x =  voom_top_sRNA$sRNA, quote = F,row.names = F, col.names = F)
#aheatmap(voom_counts)

```



```{r piRNA_information_addition}
# TODO add info for counts too
#tRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_tRNA.txt", header = FALSE, stringsAsFactors = FALSE)
lncRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_lncRNA.txt"),
                     header = FALSE, 
                     stringsAsFactors = FALSE)
unique <- read.table(file = file.path(alignments_directory, "small_sa_og_norRNA_alignments_uniquemapping_list.txt"),
                     header = FALSE, 
                     stringsAsFactors = FALSE)
piRBase <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_piRNA.txt"),
                      header = FALSE, 
                      stringsAsFactors = FALSE)
piRBaseplus15 <- read.table(file = file.path(alignments_directory, "biogenesis_alignments_norRNA_25_33/small_sa_og_norRNA_25_33_list_alignments_piRNAbaseplus15_list.txt"),
                            header = FALSE,
                            stringsAsFactors = FALSE)
YRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt"), 
                   header = FALSE, 
                   stringsAsFactors = FALSE)
snoRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_snoRNA.txt"),
                     header = FALSE, 
                     stringsAsFactors = FALSE)
miRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_miRNA.txt"),
                    header = FALSE, 
                    stringsAsFactors = FALSE, sep = "\t")
mRNA <- read.table(file = file.path(alignments_directory, "biogenesis_alignments_norRNA_25_33/small_sa_og_norRNA_25_33_list_alignments_cDNA_list.txt"),
                   header = FALSE, 
                   stringsAsFactors = FALSE)
rmsk <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker.txt"), 
                   header = FALSE, 
                   stringsAsFactors = FALSE,
                   sep = "\t")
colnames(rmsk) <- c("sRNA", "Repeat Name")
rmsk$`Repeat Name` <- gsub(pattern = "_", replacement = "", rmsk$`Repeat Name`)
rmsk_plus40 <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_plus40.txt"),
                                           header = FALSE,
                                           stringsAsFactors = FALSE)
#tRNA <- rmsk[grepl(pattern = "tRNA", x = rmsk$'Repeat Name'),]
tRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_tRNA.txt"),
                   header = FALSE,
                   stringsAsFactors = FALSE)
colnames(tRNA) <- c("sRNA", "tRNA Name")
tRNA_plus40 <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_plus40_tRNA.txt"),
                          header = FALSE, 
                          stringsAsFactors = FALSE)
colnames(tRNA_plus40) <- c("sRNA", "Extended tRNA Name")


voom.top.sRNA$'Targeting Sequence' <- strtrim(voom.top.sRNA$sRNA,25)
voom.top.sRNA$'Targeting Sequence Minus 1st 5`' <- substring(voom.top.sRNA$'Targeting Sequence',2)
voom.top.sRNA$piRBase <- ifelse(voom.top.sRNA$sRNA %in% piRBase$V1, T, F)
voom.top.sRNA$'Extended piRBase' <- ifelse(voom.top.sRNA$sRNA %in% piRBaseplus15$V1, T, F)
voom.top.sRNA$'Unique Mapping' <- ifelse(voom.top.sRNA$sRNA %in% unique$V1, T, F)
voom.top.sRNA$miRNA <- ifelse(voom.top.sRNA$sRNA %in% miRNA$V1, T, F)
voom.top.sRNA$mRNA <- ifelse(voom.top.sRNA$sRNA %in% mRNA$V1, T, F)
voom.top.sRNA$lncRNA <- ifelse(voom.top.sRNA$sRNA %in% lncRNA$V1, T, F)
voom.top.sRNA$snoRNA <- ifelse(voom.top.sRNA$sRNA %in% snoRNA$V1, T, F)
voom.top.sRNA$YRNA <- ifelse(voom.top.sRNA$sRNA %in% YRNA$V1, T, F)
voom.top.sRNA$tRNA <- ifelse(voom.top.sRNA$sRNA %in% tRNA$'sRNA', T, F)
voom.top.sRNA <- merge(voom.top.sRNA, tRNA[,c(1:2)], by = 'sRNA', all.x = T)
voom.top.sRNA$'Extended tRNA' <- ifelse(voom.top.sRNA$sRNA %in% tRNA_plus40$'sRNA', T, F)
voom.top.sRNA <- merge(voom.top.sRNA, tRNA_plus40[,c(1:2)], by = 'sRNA', all.x = T)
voom.top.sRNA$'Repeat Masker' <- ifelse(voom.top.sRNA$sRNA %in% rmsk$sRNA, T, F)
voom.top.sRNA <- merge(voom.top.sRNA, rmsk[,c(1:2)], by = 'sRNA', all.x = T)
voom.top.sRNA$'Extended Repeat Masker' <- ifelse(voom.top.sRNA$sRNA %in% rmsk_plus40$V1, T, F)


write.table(voom.top.sRNA, file = file.path(data_directory, 'stem_diff_expressions', 'GSC_vs_nonGSC_piRNA_differential_expression', 'GSC_vs_nonGSC_top_pval05_differentially_expressed_putative_piRNA.csv'),
                                            sep = ",", 
                                            row.names = F, 
                                            quote = F)

GSC_read_counts <- read_counts[rownames(read_counts) %in% voom.top.sRNA$sRNA[voom.top.sRNA$logFC > 0],]
GSC_read_counts$sRNA <- rownames(GSC_read_counts)
nonGSC_read_counts <- read_counts[rownames(read_counts) %in% voom.top.sRNA$sRNA[voom.top.sRNA$logFC < 0],]
nonGSC_read_counts$sRNA <- rownames(nonGSC_read_counts)

write.table(GSC_read_counts, file = file.path(data_directory, 'stem_diff_expressions', 'GSC_vs_nonGSC_piRNA_differential_expression',"GSC_upregulated_pval05_putative_piRNA.csv"),
            sep = ",", row.names = F,quote = F,col.names = T)
write.table(nonGSC_read_counts, file = file.path(data_directory, 'stem_diff_expressions', 'GSC_vs_nonGSC_piRNA_differential_expression', "nonGSC_upregulated_pval05_putative_piRNA.csv"),
            sep = ",", row.names = F, quote = F,col.names = T)

#annotating all reads


voom.sRNA$'Targeting Sequence' <- strtrim(voom.sRNA$sRNA,25)
voom.sRNA$'Targeting Sequence Minus 1st 5`' <- substring(voom.sRNA$'Targeting Sequence',2)
voom.sRNA$piRBase <- ifelse(voom.sRNA$sRNA %in% piRBase$V1, T, F)
voom.sRNA$'Extended piRBase' <- ifelse(voom.sRNA$sRNA %in% piRBaseplus15$V1, T, F)
voom.sRNA$'Unique Mapping' <- ifelse(voom.sRNA$sRNA %in% unique$V1, T, F)
voom.sRNA$miRNA <- ifelse(voom.sRNA$sRNA %in% miRNA$V1, T, F)
voom.sRNA$mRNA <- ifelse(voom.sRNA$sRNA %in% mRNA$V1, T, F)
voom.sRNA$lncRNA <- ifelse(voom.sRNA$sRNA %in% lncRNA$V1, T, F)
voom.sRNA$snoRNA <- ifelse(voom.sRNA$sRNA %in% snoRNA$V1, T, F)
voom.sRNA$YRNA <- ifelse(voom.sRNA$sRNA %in% YRNA$V1, T, F)
voom.sRNA$tRNA <- ifelse(voom.sRNA$sRNA %in% tRNA$'sRNA', T, F)
voom.sRNA <- merge(voom.sRNA, tRNA[,c(1:2)], by = 'sRNA', all.x = T)
voom.sRNA$'Extended tRNA' <- ifelse(voom.sRNA$sRNA %in% tRNA_plus40$'sRNA', T, F)
voom.sRNA <- merge(voom.sRNA, tRNA_plus40[,c(1:2)], by = 'sRNA', all.x = T)
voom.sRNA$'Repeat Masker' <- ifelse(voom.sRNA$sRNA %in% rmsk$sRNA, T, F)
voom.sRNA <- merge(voom.sRNA, rmsk[,c(1:2)], by = 'sRNA', all.x = T)
voom.sRNA$'Extended Repeat Masker' <- ifelse(voom.sRNA$sRNA %in% rmsk_plus40$V1, T, F)


write.table(voom.sRNA, file = file.path(data_directory, 'stem_diff_expressions', 'GSC_vs_nonGSC_piRNA_differential_expression', "GSC_vs_nonGSC_all_differentially_expressed_putative_piRNA.csv"),
            sep = ",", row.names = F, quote = F)


#tRNA derived piRNA
write.table(voom.top.sRNA$sRNA[voom.top.sRNA$tRNA], "../data/meme_files/DE_tRNA_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_tRNA_piRNA.txt"), "../data/meme_files/DE_tRNA_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_tRNA_piRNA.fa", 1, 10), "../data/meme_files/DE_tRNA_piRNA_10bp.fa" )

write.table(voom.top.sRNA$sRNA[voom.top.sRNA$mRNA], file = file.path(data_directory, "meme_files", "DE_mRNA_piRNA.txt"),
            quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta(file = file.path(data_directory, "meme_files", "DE_mRNA_piRNA.txt")),file = file.path(data_directory, "meme_files", "DE_mRNA_piRNA.fa"),
                              quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo(file = file.path(data_directory, "meme_files", "DE_mRNA_piRNA.fa"), 1, 10), file = file.path(data_directory, "meme_files", "DE_mRNA_piRNA_10bp.fa" ))

#unique derived piRNA
write.table(voom.top.sRNA$sRNA[voom.top.sRNA$unique_mapping], file = file.path(data_directory, "meme_files", "DE_UM_piRNA.txt"), quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta(file.path(data_directory, "meme_files", "DE_UM_piRNA.txt")), file = file.path(data_directory, "meme_files", "DE_UM_piRNA.fa"), quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo(file.path(data_directory, "meme_files", "DE_UM_piRNA.fa"), 1, 10), "../data/meme_files/DE_UM_piRNA_10bp.fa" )

#repeat derived piRNA
write.table(voom.top.sRNA$sRNA[voom.top.sRNA$unique_mapping == FALSE], "../data/meme_files/DE_RD_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta(file.path(data_directory, "meme_files", "DE_RD_piRNA.txt")), file.path(data_directory, "meme_files", "DE_RD_piRNA.fa"), quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo(file.path(data_directory, "meme_files", "DE_RD_piRNA.fa"), 1, 10), file.path(data_directory, "meme_files", "DE_RD_piRNA_10bp.fa"))


#repeatmasker derived piRNA
write.table(voom.top.sRNA$sRNA[voom.top.sRNA$rmsk == T & voom.top.sRNA$tRNA == F], "../data/meme_files/DE_rmsk_notRNA_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta(file.path(data_directory, "meme_files", "DE_rmsk_notRNA_piRNA.txt")), file.path(data_directory, "meme_files","DE_rmsk_notRNA_piRNA.fa"), quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo(file.path(data_directory, "meme_files", "DE_rmsk_notRNA_piRNA.fa"), 1, 10), file.path(data_directory, "meme_files","DE_rmsk_notRNA_piRNA_10bp.fa"))



```


**miRNA Differential Expression**

```{r miRNA_import_reads, echo = FALSE}

#import the desired read counts
read.counts <- read.csv(file = file.path(data_directory, "sRNA_counts", "small_sa_og_norRNA_noncRNA_18_24_counts.txt"), 
                        stringsAsFactors=FALSE,
                        row.names = 1)[,]
#select only the cell lines used in the experiment
rRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_rRNA.txt"), 
                   header = FALSE,
                   stringsAsFactors = FALSE)
read.counts <- read.counts[!(rownames(read.counts) %in% rRNA$V1),]


tRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_repeatmasker_tRNA.txt"),
                   header = FALSE,
                   stringsAsFactors = FALSE)
lncRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_lncRNA.txt"), 
                     header = FALSE, 
                     stringsAsFactors = FALSE)
unique <- read.table(file = file.path(alignments_directory, "small_sa_og_norRNA_alignments_uniquemapping_list.txt"), 
                     header = FALSE,
                     stringsAsFactors = FALSE)
piRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_piRNA.txt"),
                    header = FALSE,
                    stringsAsFactors = FALSE)
YRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_YRNA.txt"),
                   header = FALSE, 
                   stringsAsFactors = FALSE)
snoRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_snoRNA.txt"),
                     header = FALSE, 
                     stringsAsFactors = FALSE)
rmsk <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_repeatmasker.txt"), 
                   header = FALSE, 
                   stringsAsFactors = FALSE)

read.counts <- read.counts[!(rownames(read.counts) %in% tRNA$V1),]
read.counts <- read.counts[!(rownames(read.counts) %in% lncRNA$V1),]
read.counts <- read.counts[!(rownames(read.counts) %in% YRNA$V1),]
read.counts <- read.counts[!(rownames(read.counts) %in% snoRNA$V1),]
read.counts <- read.counts[!(rownames(read.counts) %in% rmsk$V1),]

# TODO redo this alignment
mycoplasm <- read.table(file = "../alignments/miRNA/small_sa_og_norRNA_noncRNA_18_24_alignments_mhyorhinisdna.txt",
                        header = FALSE, 
                        stringsAsFactors = FALSE)

read.counts <- read.counts[!(rownames(read.counts) %in% mycoplasm$V1),]


miRNA <- read.table(file = file.path(alignments_directory, "small_sa_og_norRNA_list_alignments_miRNA.txt"), 
                    header = FALSE, 
                    stringsAsFactors = FALSE, 
                    sep = "\t")
#miRNA$sequence <- sapply(miRNA$V1, FUN = strsplit, split = "\t")
#miRNA$miRNA <- unlist(sapply(miRNA$V1, FUN = strsplit, split = "\t"))[2]

read.counts <- read.counts[,4:19]
#switch the cell lines to the correct locations, control vs treatment
read.counts <- read.counts[,c(9:16, 1:8)]
#remove the low count cell lines
read.counts <- read.counts[,-c(3,11)]
#remove low count reads
read.counts <- read.counts[rowSums(cpm(read.counts) >1) >= 6,]




read.counts$sRNA <- rownames(read.counts)
read.counts$sRNA <- strtrim(read.counts$sRNA, 21)
read.counts <- merge(read.counts, 
                     miRNA, 
                     by.x = "sRNA",
                     by.y = "V1",
                     all.x = T)

putmiRNA <- 
  function(n){
    if(is.na(read.counts$V2[n]) == F){
      return(read.counts$V2[n])
    }
    else{
      return(read.counts$sRNA[n])
    }
  }
read.counts$sRNA <- sapply(X = 1:nrow(read.counts), 
                           FUN = putmiRNA)

# TODO clean this up

read.counts <- subset(read.counts, select = -V2)

read.counts <- aggregate(data = read.counts,.~sRNA, sum)

rownames(read.counts) <- read.counts$sRNA
read.counts <- subset(read.counts, select = -sRNA)


```





```{r miRNA_read_normalization, echo = FALSE}

#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge.counts <- 
  read.counts %>% 
  DGEList() %>%
  calcNormFactors(method = c("TMM"))
#voom normalization
voom.counts <- cpm(x = dge.counts,
                   log=TRUE, 
                   prior.count = 3)
#MDSplot
#plotMDS(dge.counts, labels = colnames(read.counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the counts
fit <-
  voom.counts %>%
  lmFit(design) %>%
  eBayes(trend = TRUE)


#voom fold changesmiRNA_alignments
voom.sRNA <- topTable(fit, coef = "celltypeS", sort.by="logFC", number =nrow(voom.counts) )
voom.top <- topTable(fit,coef = "celltypeS", sort.by="logFC", number =nrow(voom.counts),lfc = 0,p.value = 0.05)
#voomNormExp <- voomCounts$E
#rm(celltype);rm(design.array); rm(design); rm(dge.counts)
#my name function

#adding the reads names into its own column for merging
voom.sRNA$sRNA <- rownames(voom.sRNA)
voom.top.sRNA <- voom.top
voom.top.sRNA$sRNA <- rownames(voom.top.sRNA)

write.table(voom.sRNA, file.path(data_directory, 'stem_diff_expressions', "GSC_vs_nonGSC_miRNA_differential_expression", "GSC_vs_nonGSC_all_differentially_expressed_miRNA.csv"), sep = ",", row.names = F, quote = F)
write.table(voom.top.sRNA, file.path(data_directory, 'stem_diff_expressions',  "GSC_vs_nonGSC_miRNA_differential_expression", "GSC_vs_nonGSC_top_pval05_differentially_expressed_miRNA.csv"), sep = ",", row.names = F, quote = F)

#voom.top.sRNA$sRNA <- rownames(voom.top.sRNA)
#voom.top.sRNA <- merge(voom.sRNA, miRNA, by.x = "sRNA", by.y = "V1", all.x = T)
```






**Targeting Sequence Differential Expression**


```{r targeting_reads, echo = FALSE}

#import the desired read counts
read.counts <- read.delim(file.path(data_directory, "sRNA_counts", "small_sa_og_norRNA_25_33_counts.txt"), sep=",", stringsAsFactors=FALSE, row.names = 1)


rRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_rRNA.txt"), 
                   header = FALSE,
                   stringsAsFactors = FALSE)
read.counts <- read.counts[!(rownames(read.counts) %in% rRNA$V1),]
YRNA <- read.table(file = file.path(alignments_directory, "biogenesis_small_sa_og_norRNA", "small_sa_og_norRNA_alignments_YRNA.txt"),
                   header = FALSE, 
                   stringsAsFactors = FALSE)
read.counts <- read.counts[!(rownames(read.counts) %in% YRNA$V1),]
read.counts <- removeRepeats(read.counts)


read.counts <- read.counts[,4:19]
#sort the samples so
read.counts <- read.counts[,c(9:16, 1:8)]

read.counts <- read.counts[,-c(3,11)]


#read.counts <- read.counts[nchar(rownames(read.counts)) >= 28,]

read.counts <- read.counts[rowSums(cpm(read.counts) > 1) >= 7,]

#shorten the reads to just the predicted targeting sequence
read.counts$short <- strtrim(rownames(read.counts),25)
read.counts <- aggregate(. ~ short, data = read.counts, FUN = sum, na.rm = TRUE)
rownames(read.counts) <- read.counts$short
read.counts <- subset(read.counts, select = -short)




#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge.counts <- DGEList(counts=read.counts)
#dge.counts <- dge.counts[rowSums(cpm(dge.counts) >=1) >=7,]
dge.counts <- calcNormFactors(dge.counts, method = c("TMM"))
#voom normalization
voom.counts <- cpm(dge.counts, log = TRUE, prior.count = 3)

#MDSplot
plotMDS(voom.counts, labels = colnames(read.counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the data
fit <- 
  voom.counts %>%
  lmFit(design) %>%
  eBayes(trend = TRUE)

#sort fold changes sorted by p-value
voom.sRNA <- topTable(fit = fit, 
                      coef = "celltypeS", 
                      sort.by="logFC",
                      number =nrow(voom.counts) )
voom.top <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by="logFC",
                     number =nrow(voom.counts),
                     lfc = 2,
                     p.value = 0.05)

#adding the reads names into its own column for merging
voom.sRNA$sRNA <- rownames(voom.sRNA)
voom.top.sRNA <- voom.top
voom.top.sRNA$sRNA <- rownames(voom.top.sRNA)


#write.table(rownames(voom.sRNA), "../data/temp_read_names.txt", quote=F, row.names = F, col.names = F)
write.table(x = voom.sRNA, 
            file = file.path(data_directory, 'stem_diff_expressions', "GSC_vs_nonGSC_targeting_sequences_differential_expression", "GSC_vs_nonGSC_all_differentially_expressed_targeting_sequences.csv"),
            quote=F, 
            row.names = T, 
            col.names = T, 
            sep = ",")
write.table(x = voom.top.sRNA, file = file.path(data_directory, "GSC_vs_nonGSC_targeting_sequences_differential_expression", "GSC_vs_nonGSC_top_pval<05_differentially_expressed_targeting_sequences.csv"),
            quote=F,
            row.names = T,
            col.names = T, 
            sep = ",")


```




**SRA sRNA Differential Expression**

```{r SRA, echo = FALSE}

#import the desired read counts
read.counts <- read.csv(file = file.path(data_directory, "sRNA_counts", "sra_all_sa_og_norRNA_25_33_read.counts.txt"),
                        stringsAsFactors=FALSE, row.names = 1)[,]
#filter <- read.table("../fastx_files/sra_all_sa_og_norRNA_list.txt", header = FALSE, stringsAsFactors = FALSE)
#read.counts <- read.counts[(rownames(read.counts) %in% filter$V1),]
#read.counts <- read.counts[nchar(rownames(read.counts)) >= 25 & nchar(rownames(read.counts)) <= 33,]

colnames(read.counts) <- gsub(colnames(read.counts), 
                              pattern = "_sa_collapsed.fq", 
                              replacement = "")

#read.counts <-  read.counts[,-c(26,27,28,29,30)]
read.counts <- read.counts[,c(26:29)]

filters <- list.files(file.path(alignments_directory, "biogenesis_alignments_sra/"), full.names = T)
titles <- c("lncRNA", "miRNA", "piRNAbase", "repeat",  "snoRNA", "tRNA", "Y-RNA")

#filters

#create a dataframe of the biogenesis
df <- as.data.frame(matrix(nrow = ncol(read.counts), ncol = 0, byrow = TRUE))

df$cell_line <- colnames(read.counts)

df$total_sRNA <- colSums(read.counts)


#join all the biogenesis counts together
for(n in 1:length(filters)){
  
  #import the desired read counts
  filter <- read.table(filters[n], header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  df[,titles[n]] <- colSums(read.counts[rownames(read.counts) %in% filter$V1,])
  

}


temp <- melt(data = df[], id.vars = c(1))



ggplot(data = temp) +
#ggplot(data = temp) + 
  geom_bar(aes(y = value, x = cell_line, fill = variable), stat = "identity", position = "dodge") +
  labs(title = "SRA Hodgkin's Lymphoma piRNA Predicted Biogenesis Locations") +
  ylab("Raw Counts") +
  xlab("Cell Line") +
  scale_y_continuous(labels = comma) +
  #scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("stem", "diff")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )



temp <- melt(data = df[,-c(2)], id.vars = c(1))

#temp <- aggregate(data = temp, value ~ cell_line + cell_type+variable, sum)


ggplot(data = temp ) +
#ggplot(data = temp) + 
  geom_boxplot(aes(y = value, x = variable, fill = cell_line)) +
  labs(title = "Biogenesis of Reads") +
  ylab("Counts Per Million") +
  xlab("Biogenesis Location") +
  scale_y_continuous(labels = comma) +
  scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("GSC", "Non-GSC")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank() 
  )






#import the desired read counts
read.counts <- read.delim(file.path(data_directory, 'sRNA_counts', "SRA_clusters_counts.txt"), sep=",", stringsAsFactors=FALSE, row.names = 1)[,]
#select only the cell lines used in the experiment
#read.counts <- read.counts[nchar(rownames(read.counts)) >17,]

read.counts <- read.counts[,c(2,3,4,5,6,7,1,8,9,10)]

colnames(read.counts) <- c( "OA1", "OA2", "O11", "O12", "O21", "O22","T1", "T2", "T3", "T4" )


read.counts <- read.counts[nchar(rownames(read.counts)) >= 25 & nchar(rownames(read.counts)) <=33, ]

filters <- list.files(file.path(data_directory, 'sRNA_counts', "biogenesis_small_sa_og_norRNA/"), full.names = T)
titles <- c("lncRNA", "miRNA", "piRNAbase", "rRNA",  "snoRNA", "tRNA", "Y-RNA")


#create a dataframe of the biogenesis
df <- as.data.frame(matrix(nrow = 10, ncol = 0, byrow = TRUE))

df$cell_line <- c( "OA1", "OA2", "O11", "O12", "O21", "O22","T1", "T2", "T3", "T4" )
df$cell_type <- factor(c(rep("Ovary", 6), (rep("Testis",4))))

df$total_sRNA <- colSums(read.counts)


#join all the biogenesis counts together
for(n in 1:length(filters)){
  
  #import the desired read counts
  filter <- read.table(filters[n], header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  df[,titles[n]] <- colSums(read.counts[rownames(read.counts) %in% filter$V1,])
  

}


temp <- melt(data = df[,-c(2)], id.vars = c(1))



ggplot(data = temp) +
#ggplot(data = temp) + 
  geom_bar(aes(y = value, x = cell_line, fill = variable), stat = "identity", position = "dodge") +
  labs(title = "Read Counts from Predicted Biogenesis Locations") +
  ylab("Raw Counts") +
  xlab("Cell Line") +
  scale_y_continuous(labels = comma) +
  #scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("stem", "diff")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )



```

