---
title: "sRNA_differential_expression"
author: "Andrew Burr"
date: "April 23, 2017"
output: html_document
---



```{r setup, include=FALSE}
source('global.R')
#force R to not use scientific notation
options(scipen=999)




#load in the gene expression from RNA seq experiment
gene_expression <- read.csv("../../RNAseq/data/StemVsDiff_all_DE_genes.txt", sep = ",", stringsAsFactors = F)
transcript_expression <- read.csv("../../RNAseq/data/StemVsDiff_all_DE_transcripts.csv", sep = ",", stringsAsFactors = F)
#load in the differentially expressed genes info from the RNA seq experiment
top_genes <- read.csv("../../RNAseq/data/StemVsDiff_top_DE_genes.txt", stringsAsFactors=FALSE)

conversion <- read.table("../../conversion_sets//ensembl_conversion.txt", header = T, stringsAsFactors = F, sep = "\t")



```


**piRNA Differential Expression**

```{r piRNA_differential_expression, echo = FALSE}

#import the desired read counts
read_counts <- read.delim("../sRNA_counts/small_sa_og_norRNA_25_33_counts.txt", sep=",", stringsAsFactors=FALSE, row.names = 1)[,]

#remove non used reads
rRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_rRNA.txt", header = FALSE, stringsAsFactors = FALSE)
read_counts <- read_counts[!(rownames(read_counts) %in% rRNA$V1),]
YRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt", header = FALSE, stringsAsFactors = FALSE)
read_counts <- read_counts[!(rownames(read_counts) %in% YRNA$V1),]

#keep only cell lines from experiment
read_counts <- read_counts[,4:19]
#switch the cell lines
read_counts <- read_counts[,c(9:16, 1:8)]

read_counts <- read_counts[,-c(3,11)]

read_counts <- read_counts[!rowSums(read_counts) == 0,]

#read_counts <- removeRepeats(read_counts)
#remove low counts for differential expression
read_counts <- read_counts[rowSums(cpm(read_counts) >= 1) >= 7,]
read_counts <- read_counts[!rowSums(read_counts) == 0,]




#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge_counts <- DGEList(counts=read_counts)
#dge_counts <- dge_counts[rowSums(cpm(dge_counts) >=1) >=7,]
dge_counts <- calcNormFactors(dge_counts, method = c("TMM"))

#voom normalization
voom_counts <- cpm(dge_counts,log=TRUE, prior.count = 3)
#MDSplot
plotMDS(dge_counts, labels = colnames(read_counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the counts
fit <- eBayes(lmFit(voom_counts,design), trend = TRUE)


#voom fold changesmiRNA_alignments
voom_sRNA <- topTable(fit, coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts) )
voom_top <- topTable(fit,coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts),lfc = 0,p.value = 0.05)
#voomNormExp <- voomCounts$E
#rm(celltype);rm(design.array); rm(design); rm(dge_counts)
#my name function

#adding the reads names into its own column for merging
voom_sRNA$sRNA <- rownames(voom_sRNA)
voom_top_sRNA <- voom_top
voom_top_sRNA$sRNA <- rownames(voom_top_sRNA)

#voom_top_sRNA$sRNA <- strtrim(voom_top_sRNA$sRNA, 25)
#write.table("../fastx_files/DE_piRNA_list.txt",x =  voom_top_sRNA$sRNA, quote = F,row.names = F, col.names = F)
#aheatmap(voom_counts)

```



```{r piRNA_information_addition}

#tRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_tRNA.txt", header = FALSE, stringsAsFactors = FALSE)
lncRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_lncRNA.txt", header = FALSE, stringsAsFactors = FALSE)
unique <- read.table("../alignments/small_sa_og_norRNA_alignments_uniquemapping_list.txt", header = FALSE, stringsAsFactors = FALSE)
piRBase <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_piRNA.txt", header = FALSE, stringsAsFactors = FALSE)
piRBaseplus15 <- read.table("../alignments/biogenesis_alignments_norRNA_25_33/small_sa_og_norRNA_25_33_list_alignments_piRNAbaseplus15_list.txt", header = FALSE, stringsAsFactors = FALSE)
YRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt", header = FALSE, stringsAsFactors = FALSE)
snoRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_snoRNA.txt", header = FALSE, stringsAsFactors = FALSE)
miRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_miRNA.txt", header = FALSE, stringsAsFactors = FALSE, sep = "\t")
mRNA <- read.table("../alignments/biogenesis_alignments_norRNA_25_33/small_sa_og_norRNA_25_33_list_alignments_cDNA_list.txt", header = FALSE, stringsAsFactors = FALSE)
rmsk <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker.txt", header = FALSE, stringsAsFactors = FALSE, sep = "\t")
colnames(rmsk) <- c("sRNA", "Repeat Name")
rmsk$`Repeat Name` <- gsub(pattern = "_", replacement = "", rmsk$`Repeat Name`)
rmsk_plus40 <- read.table("../alignments/small_sa_og_norRNA_alignments_repeatmasker_plus40.txt", header = FALSE, stringsAsFactors = FALSE)
#tRNA <- rmsk[grepl(pattern = "tRNA", x = rmsk$'Repeat Name'),]
tRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_tRNA.txt", header = FALSE, stringsAsFactors = FALSE)
colnames(tRNA) <- c("sRNA", "tRNA Name")
tRNA_plus40 <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_plus40_tRNA.txt", header = FALSE, stringsAsFactors = FALSE)
colnames(tRNA_plus40) <- c("sRNA", "Extended tRNA Name")


voom_top_sRNA$'Targeting Sequence' <- strtrim(voom_top_sRNA$sRNA,25)
voom_top_sRNA$'Targeting Sequence Minus 1st 5`' <- substring(voom_top_sRNA$'Targeting Sequence',2)
voom_top_sRNA$piRBase <- ifelse(voom_top_sRNA$sRNA %in% piRBase$V1, T, F)
voom_top_sRNA$'Extended piRBase' <- ifelse(voom_top_sRNA$sRNA %in% piRBaseplus15$V1, T, F)
voom_top_sRNA$'Unique Mapping' <- ifelse(voom_top_sRNA$sRNA %in% unique$V1, T, F)
voom_top_sRNA$miRNA <- ifelse(voom_top_sRNA$sRNA %in% miRNA$V1, T, F)
voom_top_sRNA$mRNA <- ifelse(voom_top_sRNA$sRNA %in% mRNA$V1, T, F)
voom_top_sRNA$lncRNA <- ifelse(voom_top_sRNA$sRNA %in% lncRNA$V1, T, F)
voom_top_sRNA$snoRNA <- ifelse(voom_top_sRNA$sRNA %in% snoRNA$V1, T, F)
voom_top_sRNA$YRNA <- ifelse(voom_top_sRNA$sRNA %in% YRNA$V1, T, F)
voom_top_sRNA$tRNA <- ifelse(voom_top_sRNA$sRNA %in% tRNA$'sRNA', T, F)
voom_top_sRNA <- merge(voom_top_sRNA, tRNA[,c(1:2)], by = 'sRNA', all.x = T)
voom_top_sRNA$'Extended tRNA' <- ifelse(voom_top_sRNA$sRNA %in% tRNA_plus40$'sRNA', T, F)
voom_top_sRNA <- merge(voom_top_sRNA, tRNA_plus40[,c(1:2)], by = 'sRNA', all.x = T)
voom_top_sRNA$'Repeat Masker' <- ifelse(voom_top_sRNA$sRNA %in% rmsk$sRNA, T, F)
voom_top_sRNA <- merge(voom_top_sRNA, rmsk[,c(1:2)], by = 'sRNA', all.x = T)
voom_top_sRNA$'Extended Repeat Masker' <- ifelse(voom_top_sRNA$sRNA %in% rmsk_plus40$V1, T, F)


write.table(voom_top_sRNA, "../data/differential_expression//GSC_vs_nonGSC_top_pval<05_differentially_expressed_putative_piRNA.csv", sep = ",", row.names = F, quote = F)

GSC_read_counts <- read_counts[rownames(read_counts) %in% voom_top_sRNA$sRNA[voom_top_sRNA$logFC > 0],]
GSC_read_counts$sRNA <- rownames(GSC_read_counts)
nonGSC_read_counts <- read_counts[rownames(read_counts) %in% voom_top_sRNA$sRNA[voom_top_sRNA$logFC < 0],]
nonGSC_read_counts$sRNA <- rownames(nonGSC_read_counts)

write.table(GSC_read_counts, "../data/GSC_upregulated_pval<05_putative_piRNA.csv", sep = ",", row.names = F, quote = F,col.names = T)
write.table(nonGSC_read_counts, "../data/nonGSC_upregulated_pval<05_putative_piRNA.csv", sep = ",", row.names = F, quote = F,col.names = T)

#annotating all reads


voom_sRNA$'Targeting Sequence' <- strtrim(voom_sRNA$sRNA,25)
voom_sRNA$'Targeting Sequence Minus 1st 5`' <- substring(voom_sRNA$'Targeting Sequence',2)
voom_sRNA$piRBase <- ifelse(voom_sRNA$sRNA %in% piRBase$V1, T, F)
voom_sRNA$'Extended piRBase' <- ifelse(voom_sRNA$sRNA %in% piRBaseplus15$V1, T, F)
voom_sRNA$'Unique Mapping' <- ifelse(voom_sRNA$sRNA %in% unique$V1, T, F)
voom_sRNA$miRNA <- ifelse(voom_sRNA$sRNA %in% miRNA$V1, T, F)
voom_sRNA$mRNA <- ifelse(voom_sRNA$sRNA %in% mRNA$V1, T, F)
voom_sRNA$lncRNA <- ifelse(voom_sRNA$sRNA %in% lncRNA$V1, T, F)
voom_sRNA$snoRNA <- ifelse(voom_sRNA$sRNA %in% snoRNA$V1, T, F)
voom_sRNA$YRNA <- ifelse(voom_sRNA$sRNA %in% YRNA$V1, T, F)
voom_sRNA$tRNA <- ifelse(voom_sRNA$sRNA %in% tRNA$'sRNA', T, F)
voom_sRNA <- merge(voom_sRNA, tRNA[,c(1:2)], by = 'sRNA', all.x = T)
voom_sRNA$'Extended tRNA' <- ifelse(voom_sRNA$sRNA %in% tRNA_plus40$'sRNA', T, F)
voom_sRNA <- merge(voom_sRNA, tRNA_plus40[,c(1:2)], by = 'sRNA', all.x = T)
voom_sRNA$'Repeat Masker' <- ifelse(voom_sRNA$sRNA %in% rmsk$sRNA, T, F)
voom_sRNA <- merge(voom_sRNA, rmsk[,c(1:2)], by = 'sRNA', all.x = T)
voom_sRNA$'Extended Repeat Masker' <- ifelse(voom_sRNA$sRNA %in% rmsk_plus40$V1, T, F)


write.table(voom_sRNA, "../data/differential_expression/GSC_vs_nonGSC_all_differentially_expressed_putative_piRNA.csv", sep = ",", row.names = F, quote = F)


#tRNA derived piRNA
write.table(voom_top_sRNA$sRNA[voom_top_sRNA$tRNA], "../data/meme_files/DE_tRNA_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_tRNA_piRNA.txt"), "../data/meme_files/DE_tRNA_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_tRNA_piRNA.fa", 1, 10), "../data/meme_files/DE_tRNA_piRNA_10bp.fa" )

write.table(voom_top_sRNA$sRNA[voom_top_sRNA$mRNA], "../data/meme_files/DE_mRNA_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_mRNA_piRNA.txt"), "../data/meme_files/DE_mRNA_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_mRNA_piRNA.fa", 1, 10), "../data/meme_files/DE_mRNA_piRNA_10bp.fa" )

#unique derived piRNA
write.table(voom_top_sRNA$sRNA[voom_top_sRNA$unique_mapping], "../data/meme_files/DE_UM_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_UM_piRNA.txt"), "../data/meme_files/DE_UM_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_UM_piRNA.fa", 1, 10), "../data/meme_files/DE_UM_piRNA_10bp.fa" )

#repeat derived piRNA
write.table(voom_top_sRNA$sRNA[voom_top_sRNA$unique_mapping == FALSE], "../data/meme_files/DE_RD_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_RD_piRNA.txt"), "../data/meme_files/DE_RD_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_RD_piRNA.fa", 1, 10), "../data/meme_files/DE_RD_piRNA_10bp.fa" )


#repeatmasker derived piRNA
write.table(voom_top_sRNA$sRNA[voom_top_sRNA$rmsk == T & voom_top_sRNA$tRNA == F], "../data/meme_files/DE_rmsk_notRNA_piRNA.txt", quote=F, row.names = F, col.names = F)
write.table(createpiRNAFasta("../data/meme_files/DE_rmsk_notRNA_piRNA.txt"), "../data/meme_files/DE_rmsk_notRNA_piRNA.fa", quote= F, row.names = F, col.names = F, sep = "\t")
writeXStringSet(trimFastaTo("../data/meme_files/DE_rmsk_notRNA_piRNA.fa", 1, 10), "../data/meme_files/DE_rmsk_notRNA_piRNA_10bp.fa" )



```


**miRNA Differential Expression**

```{r miRNA_import_reads, echo = FALSE}

#import the desired read counts
read_counts <- read.delim("../sRNA_counts/small_sa_og_norRNA_noncRNA_18_24_counts.txt", sep=",", stringsAsFactors=FALSE, row.names = 1)[,]
#select only the cell lines used in the experiment
rRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_rRNA.txt", header = FALSE, stringsAsFactors = FALSE)
read_counts <- read_counts[!(rownames(read_counts) %in% rRNA$V1),]


tRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker_tRNA.txt", header = FALSE, stringsAsFactors = FALSE)
lncRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_lncRNA.txt", header = FALSE, stringsAsFactors = FALSE)
unique <- read.table("../alignments/small_sa_og_norRNA_alignments_uniquemapping_list.txt", header = FALSE, stringsAsFactors = FALSE)
piRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_piRNA.txt", header = FALSE, stringsAsFactors = FALSE)
YRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt", header = FALSE, stringsAsFactors = FALSE)
snoRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_snoRNA.txt", header = FALSE, stringsAsFactors = FALSE)
rmsk <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_repeatmasker.txt", header = FALSE, stringsAsFactors = FALSE)

read_counts <- read_counts[!(rownames(read_counts) %in% tRNA$V1),]
read_counts <- read_counts[!(rownames(read_counts) %in% lncRNA$V1),]
read_counts <- read_counts[!(rownames(read_counts) %in% YRNA$V1),]
read_counts <- read_counts[!(rownames(read_counts) %in% snoRNA$V1),]
read_counts <- read_counts[!(rownames(read_counts) %in% rmsk$V1),]

mycoplasm <- read.table("../alignments/miRNA/small_sa_og_norRNA_noncRNA_18_24_alignments_mhyorhinisdna.txt", header = FALSE, stringsAsFactors = FALSE)

read_counts <- read_counts[!(rownames(read_counts) %in% mycoplasm$V1),]


miRNA <- read.table("../alignments/small_sa_og_norRNA_list_alignments_miRNA.txt", header = FALSE, stringsAsFactors = FALSE, sep = "\t")
#miRNA$sequence <- sapply(miRNA$V1, FUN = strsplit, split = "\t")
#miRNA$miRNA <- unlist(sapply(miRNA$V1, FUN = strsplit, split = "\t"))[2]

read_counts <- read_counts[,4:19]
#switch the cell lines to the correct locations, control vs treatment
read_counts <- read_counts[,c(9:16, 1:8)]
#remove the low count cell lines
read_counts <- read_counts[,-c(3,11)]
#remove low count reads
read_counts <- read_counts[rowSums(cpm(read_counts) >1) >= 6,]




read_counts$sRNA <- rownames(read_counts)
read_counts$sRNA <- strtrim(read_counts$sRNA, 21)
read_counts <- merge(read_counts, miRNA, by.x = "sRNA", by.y = "V1", all.x = T)

putmiRNA <- function(n){
  if(is.na(read_counts$V2[n]) == F){
    return(read_counts$V2[n])
  }
  else{
    return(read_counts$sRNA[n])
  }
}
read_counts$sRNA <- sapply(X = 1:nrow(read_counts), FUN = putmiRNA)


read_counts <- subset(read_counts, select = -V2)

read_counts <- aggregate(data = read_counts,.~sRNA, sum)

rownames(read_counts) <- read_counts$sRNA
read_counts <- subset(read_counts, select = -sRNA)


```





```{r miRNA_read_normalization, echo = FALSE}

#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge_counts <- DGEList(counts=read_counts)
#dge_counts <- dge_counts[rowSums(cpm(dge_counts) >=1) >=7,]
dge_counts <- calcNormFactors(dge_counts, method = c("TMM"))

#voom normalization
voom_counts <- cpm(dge_counts,log=TRUE, prior.count = 3)
#MDSplot
#plotMDS(dge_counts, labels = colnames(read_counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the counts
fit <- eBayes(lmFit(voom_counts,design), trend = TRUE)


#voom fold changesmiRNA_alignments
voom_sRNA <- topTable(fit, coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts) )
voom_top <- topTable(fit,coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts),lfc = 0,p.value = 0.05)
#voomNormExp <- voomCounts$E
#rm(celltype);rm(design.array); rm(design); rm(dge_counts)
#my name function

#adding the reads names into its own column for merging
voom_sRNA$sRNA <- rownames(voom_sRNA)
voom_top_sRNA <- voom_top
voom_top_sRNA$sRNA <- rownames(voom_top_sRNA)

write.table(voom_sRNA, "../data/GSC_vs_nonGSC_miRNA_differential_expression//GSC_vs_nonGSC_all_differentially_expressed_miRNA.csv", sep = ",", row.names = F, quote = F)
write.table(voom_top_sRNA, "../data/GSC_vs_nonGSC_miRNA_differential_expression/GSC_vs_nonGSC_top_pval<05_differentially_expressed_miRNA.csv", sep = ",", row.names = F, quote = F)

#voom_top_sRNA$sRNA <- rownames(voom_top_sRNA)
#voom_top_sRNA <- merge(voom_sRNA, miRNA, by.x = "sRNA", by.y = "V1", all.x = T)
```






**Targeting Sequence Differential Expression**


```{r targeting_reads, echo = FALSE}

#import the desired read counts
read_counts <- read.delim("../sRNA_counts/small_sa_og_norRNA_25_33_counts.txt", sep=",", stringsAsFactors=FALSE, row.names = 1)

rRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_rRNA.txt", header = FALSE, stringsAsFactors = FALSE)
read_counts <- read_counts[!(rownames(read_counts) %in% rRNA$V1),]
YRNA <- read.table("../alignments/biogenesis_small_sa_og_norRNA/small_sa_og_norRNA_alignments_YRNA.txt", header = FALSE, stringsAsFactors = FALSE)
read_counts <- read_counts[!(rownames(read_counts) %in% YRNA$V1),]
read_counts <- removeRepeats(read_counts)


read_counts <- read_counts[,4:19]
#sort the samples so
read_counts <- read_counts[,c(9:16, 1:8)]

read_counts <- read_counts[,-c(3,11)]


#read_counts <- read_counts[nchar(rownames(read_counts)) >= 28,]

read_counts <- read_counts[rowSums(cpm(read_counts) > 1) >= 7,]

#shorten the reads to just the predicted targeting sequence
read_counts$short <- strtrim(rownames(read_counts),25)
read_counts <- aggregate(. ~ short, data = read_counts, FUN = sum, na.rm = TRUE)
rownames(read_counts) <- read_counts$short
read_counts <- subset(read_counts, select = -short)




#factor for the cell tyoe
celltype <- factor(c(rep("D",7),  #D is differentiated of non-GSC
                     rep("S",7)), #S is stem or GSC
                   levels = c("D", "S"))
#factor for the paired samples
cellline <- factor(c(1:7, 1:7), levels = 1:7)
#design matix of the experiment
design <-model.matrix(~cellline + celltype)

#converting to DGEList and keeping only expressions significantly above 0
dge_counts <- DGEList(counts=read_counts)
#dge_counts <- dge_counts[rowSums(cpm(dge_counts) >=1) >=7,]
dge_counts <- calcNormFactors(dge_counts, method = c("TMM"))
#voom normalization
voom_counts <- cpm(dge_counts, log = TRUE, prior.count = 3)

#MDSplot
plotMDS(voom_counts, labels = colnames(read_counts), cex = 1, main = "MDS Plot for piRNA Expressions", dim.plot = c(1,2))

#create the linear fit for the data
fit <- eBayes(lmFit(voom_counts,design), trend = TRUE)

#sort fold changes sorted by p-value
voom_sRNA <- topTable(fit, coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts) )
voom_top <- topTable(fit,coef = "celltypeS", sort.by="logFC", number =nrow(voom_counts),lfc = 2,p.value = 0.05)

#adding the reads names into its own column for merging
voom_sRNA$sRNA <- rownames(voom_sRNA)
voom_top_sRNA <- voom_top
voom_top_sRNA$sRNA <- rownames(voom_top_sRNA)


#write.table(rownames(voom_sRNA), "../data/temp_read_names.txt", quote=F, row.names = F, col.names = F)
write.table(voom_sRNA, "../data/GSC_vs_nonGSC_targeting_sequences_differential_expression/GSC_vs_nonGSC_all_differentially_expressed_targeting_sequences.csv", quote=F, row.names = T, col.names = T, sep = ",")
write.table(voom_top_sRNA, "../data/GSC_vs_nonGSC_targeting_sequences_differential_expression/GSC_vs_nonGSC_top_pval<05_differentially_expressed_targeting_sequences.csv", quote=F, row.names = T, col.names = T, sep = ",")


```




**SRA sRNA Differential Expression**

```{r SRA, echo = FALSE}

#import the desired read counts
read_counts <- read.delim("../sRNA_counts/sra_all_sa_og_norRNA_25_33_read_counts.txt", sep=",", stringsAsFactors=FALSE, row.names = 1)[,]
#filter <- read.table("../fastx_files/sra_all_sa_og_norRNA_list.txt", header = FALSE, stringsAsFactors = FALSE)
#read_counts <- read_counts[(rownames(read_counts) %in% filter$V1),]
#read_counts <- read_counts[nchar(rownames(read_counts)) >= 25 & nchar(rownames(read_counts)) <= 33,]

colnames(read_counts) <- gsub(colnames(read_counts), pattern = "_sa_collapsed.fq", replacement = "")

#read_counts <-  read_counts[,-c(26,27,28,29,30)]
read_counts <- read_counts[,c(26:29)]

filters <- list.files("../alignments/biogenesis_alignments_sra/", full.names = T)
titles <- c("lncRNA", "miRNA", "piRNAbase", "repeat",  "snoRNA", "tRNA", "Y-RNA")

#filters

#create a dataframe of the biogenesis
df <- as.data.frame(matrix(nrow = ncol(read_counts), ncol = 0, byrow = TRUE))

df$cell_line <- colnames(read_counts)

df$total_sRNA <- colSums(read_counts)


#join all the biogenesis counts together
for(n in 1:length(filters)){
  
  #import the desired read counts
  filter <- read.table(filters[n], header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  df[,titles[n]] <- colSums(read_counts[rownames(read_counts) %in% filter$V1,])
  

}


temp <- melt(data = df[], id.vars = c(1))



ggplot(data = temp) +
#ggplot(data = temp) + 
  geom_bar(aes(y = value, x = cell_line, fill = variable), stat = "identity", position = "dodge") +
  labs(title = "SRA Hodgkin's Lymphoma piRNA Predicted Biogenesis Locations") +
  ylab("Raw Counts") +
  xlab("Cell Line") +
  scale_y_continuous(labels = comma) +
  #scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("stem", "diff")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )



temp <- melt(data = df[,-c(2)], id.vars = c(1))

#temp <- aggregate(data = temp, value ~ cell_line + cell_type+variable, sum)


ggplot(data = temp ) +
#ggplot(data = temp) + 
  geom_boxplot(aes(y = value, x = variable, fill = cell_line)) +
  labs(title = "Biogenesis of Reads") +
  ylab("Counts Per Million") +
  xlab("Biogenesis Location") +
  scale_y_continuous(labels = comma) +
  scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("GSC", "Non-GSC")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank() 
  )






#import the desired read counts
read_counts <- read.delim("../sRNA_counts/SRA_clusters_counts.txt", sep=",", stringsAsFactors=FALSE, row.names = 1)[,]
#select only the cell lines used in the experiment
#read_counts <- read_counts[nchar(rownames(read_counts)) >17,]

read_counts <- read_counts[,c(2,3,4,5,6,7,1,8,9,10)]

colnames(read_counts) <- c( "OA1", "OA2", "O11", "O12", "O21", "O22","T1", "T2", "T3", "T4" )


read_counts <- read_counts[nchar(rownames(read_counts)) >= 25 & nchar(rownames(read_counts)) <=33, ]

filters <- list.files("../alignments/biogenesis_small_sa_og_norRNA/", full.names = T)
titles <- c("lncRNA", "miRNA", "piRNAbase", "rRNA",  "snoRNA", "tRNA", "Y-RNA")


#create a dataframe of the biogenesis
df <- as.data.frame(matrix(nrow = 10, ncol = 0, byrow = TRUE))

df$cell_line <- c( "OA1", "OA2", "O11", "O12", "O21", "O22","T1", "T2", "T3", "T4" )
df$cell_type <- factor(c(rep("Ovary", 6), (rep("Testis",4))))

df$total_sRNA <- colSums(read_counts)


#join all the biogenesis counts together
for(n in 1:length(filters)){
  
  #import the desired read counts
  filter <- read.table(filters[n], header = FALSE, stringsAsFactors = FALSE, sep = "\t")
  df[,titles[n]] <- colSums(read_counts[rownames(read_counts) %in% filter$V1,])
  

}


temp <- melt(data = df[,-c(2)], id.vars = c(1))



ggplot(data = temp) +
#ggplot(data = temp) + 
  geom_bar(aes(y = value, x = cell_line, fill = variable), stat = "identity", position = "dodge") +
  labs(title = "Read Counts from Predicted Biogenesis Locations") +
  ylab("Raw Counts") +
  xlab("Cell Line") +
  scale_y_continuous(labels = comma) +
  #scale_fill_brewer(name = "Phenotype", labels = c("GSC", "Non-GSC"), breaks = c("stem", "diff")) +
  
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 0),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )



```

