---
title: "GSC RNAseq"
author: "Gurkan Bebek & Andrew Burr"
date: "December 4, 2015"
output: html_document
---

This is RNAseq analysis of the comparison between GSC vs differentiated NPC (10 vs. 10)  


```{r, cache=TRUE,echo=FALSE,eval=TRUE,warning=FALSE}



source('global.R')
source('RNAseq_functions.R')

conversion <- read.table("../../conversion_sets/ensembl_conversion.txt", header = T)



```




```{r GSC_vs_nonGSC}

# load in the star counts
read.counts <-importStarCounts("../star_v5/", column = 4)
# removing uneeded text from column names
colnames(read.counts) <- gsub(pattern = "_dna_ReadsPerGene.out.tab", 
                              "",
                              colnames(read.counts))


# change column names to the samples names form the data
# if wish to be viewed this way
colnames(read.counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++",
                           "387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", 
                           "1123++", "528++", 
                           "1123-", "528-", 
                           "hNP1_old","hNP1_new","17231","16157")
# write the read counts for later reference
write.table(x = read.counts,
            file = "../data/all_read.counts.txt",
            sep = "\t",
            col.names = T)


#remove reads that did not get counted due to multi mapping and unalignment
read.counts <- read.counts[!rownames(read.counts)=="N_unmapped",]
read.counts <- read.counts[!rownames(read.counts)=="N_multimapping",]
read.counts <- read.counts[!rownames(read.counts)=="N_noFeature",]
read.counts <- read.counts[!rownames(read.counts)=="N_ambiguous",]


# remove unnecesary cell lines
read.counts <- subset(read.counts,
                      select = -c(`1123++`, `528++`, 
                                  `1123-`,`528-`, 
                                  hNP1_old, hNP1_new, 
                                  `17231`, `16157` ))

# write this experiments counts to a file
write.table(x = read.counts, 
            file = "../data/GSC_and_nonGSC_read.counts.txt",
            sep = "\t",
            col.names = T)
# organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
read.counts <- read.counts[,c(9:16, 1:8)]

# factor of the cell types
celltype <- factor(c(rep("D",8),   #D stands for differentiated cell or non-GSCs
                     rep("S",8)),  #S stands for stem cell of GSC
                   levels = c("D","S"))  

cellline <- factor(c(1:8, 1:8))  #Factor for the samples used. 8 paired samples
# create the design matrix based on the factors above
design <- model.matrix(~ cellline + celltype)   


# removing low count genes
# convert to DGE List
# Normalize
dge_counts <- 
  read.counts[rowSums(cpm(read.counts) >= 1) >= 8,] %>%
  DGEList(counts = ) %>%
  calcNormFactors(method = c("TMM"))



#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <- voom(counts = dge_counts, 
                    design = design,
                    plot=TRUE)
#MDS plot of the data
plotMDS(x = dge_counts, 
        cex = 1,
        main = "MDS of Stem vs Diff",
        labels = colnames(read.counts),
        top = 500, 
        dim.plot = c(1,2),
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <-
  eBayes(
    lmFit(object = voom_counts,
          design = design)
    )

#voom fold changes and p values for all genes
voom_all <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by ="p",
                     number = nrow(voom_counts)
                     )
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit = it,
                     coef = "celltypeS",
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_all_HGNC <- EnsembltoHGNC(voom_all)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


#saving the  top tables for use in other scripts
#commented out for use only when needed

write.table(x = voom_all_HGNC, 
            file = "../data/GSC_vs_nonGSC_all_differentially_expressed_genes.csv",
            quote = FALSE, 
            sep = ",", 
            row.names = T)
write.table(x = voom_top_HGNC,
            file = "../data/GSC_vs_nonGSC_top_pval05_differentially_expressed_genes.csv", 
            quote = FALSE, 
            sep = ",",
            row.names = T)
#write.table(x = as.data.frame(voom_counts$E), "../data/star_v5_normalized_counts.txt", quote = FALSE,  col.names = TRUE, sep = ",")



```


```{r fgsea}
# read in the conversion file to change gene identifiers
conversion <- read.table(file = "../../conversion_sets/ensembl_to_entrez.txt", 
                         stringsAsFactors = F, 
                         sep = "\t", 
                         header = T)
stats <- as.data.frame(voom_all_HGNC$logFC)
stats$hgnc_symbol <- voom_all_HGNC$hgnc_symbol
#stats <- stats[stats$symbol != "",]
stats <- merge(stats,conversion, by = 'hgnc_symbol')
ranks <- stats$`voom_all_HGNC$logFC`
stats$entrez_gene_id <- as.character(stats$entrez_gene_id)
names(ranks) <- stats$entrez_gene_id
ranks <- sort(ranks)

pathways <- reactomePathways(stats$entrez_gene_id)
test <- fgsea(pathways,exampleRanks)

data("examplePathways")
data("exampleRanks")
fgseaRes <- fgsea(pathways = pathways,
                  stats = ranks, 
                  nperm = 10000,
                  maxSize = 500)


pdf("../GSEA/GSC_vs_nonGSC_gsea.pdf", 14,8)
topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
print(
  
  plotGseaTable(pathways[topPathways], 
                ranks, 
                fgseaRes, 
                gseaParam = 0.5))
dev.off()




```




```{r GSC_vs_HNPC}

#load in the star counts
read.counts <- importStarCounts("../star_v5/", column = 4)

# write the read counts for later reference
write.table(x = read.counts,
            file = "../data/GSC_vs_HNPC_read.counts.txt",
            sep = "\t",
            col.names = T)


# remove reads that did not get counted due to multi mapping and unalignment
read.counts <- read.counts[!rownames(read.counts)=="N_unmapped",]
read.counts <- read.counts[!rownames(read.counts)=="N_multimapping",]
read.counts <- read.counts[!rownames(read.counts)=="N_noFeature",]
read.counts <- read.counts[!rownames(read.counts)=="N_ambiguous",]


# change column names to the samples names form the data
colnames(read.counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++",
                           "387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", 
                           "1123++", "528++", 
                           "1123-", "528-",
                           "hNP1_old","hNP1_new","17231","16157")

# remove unneccessary cell lines
read.counts <- subset(read.counts, 
                      select = -c(`387-`,`456-`, `3359-`,`3691-`,`3832-`,`4121-`,`4302-`,`H2S-`,
                                  `1123++`, `528++`, `1123-`, `528-`,`17231`,`16157`))


#organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
read.counts <- read.counts[,c(9,10,1:8) ]
#factor of the cell types
celltype <- factor(c(rep("NPC",2),
                     rep("GSC",8)), 
                   levels = c("NPC","GSC"))
#cellline <- factor(c(1:8, 1:8, 9,9))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above
design <- model.matrix(~ celltype)   

#Voom recommends reads with low counts be removed
read.counts <- 
# converting to DGEList
# removing low count genes
dge_counts <- 
  read.counts[rowSums(cpm(read.counts) >= 1) >= 2,] %>%
  DGEList(counts = read.counts)

# normalization
dge_counts <- calcNormFactors(dge_counts, 
                              method = c("TMM"))

# voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(counts = dge_counts, 
                   design = design, 
                   plot = TRUE)
#MDS plot of the data
plotMDS(x = dge_counts,
        cex = 1,
        main = "MDS of Stem vs Diff",
        labels = colnames(read.counts),
        top = 10000, 
        dim.plot = c(1,2), 
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <- 
  eBayes(
    lmFit(object = voom_counts,
          design = design))

#voom fold changes and p values for all genes
voom_all_genes <- topTable(fit = fit, 
                           coef = "celltypeGSC",
                           sort.by="p",
                           number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit = fit,
                     coef = "celltypeGSC",
                     sort.by = "p",
                     number = nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_all_genes_HGNC <- EnsembltoHGNC(voom_all_genes)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


#saving the  top tables for use in other scripts
#commented out for use only when needed


# write expression results to file
write.table(x = voom_all_HGNC, 
            file = "../data/GSC_vs_HNPC_all_differentially_expressed_genes.csv",
            quote = FALSE,
            sep = ",",
            row.names = F)
write.table(x = voom_top_HGNC, 
            file = "../data/GSC_vs_HNPC_top_pval05_differentially_expressed_genes.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)


# load conversion file to convert gene names
conversion <- read.table(file = "../../conversion_sets/ensembl_to_entrez.txt",
                         stringsAsFactors = F, 
                         sep = "\t",
                         header = T)
stats <- as.data.frame(voom_all_HGNC$logFC)
stats$hgnc_symbol <- voom_all_HGNC$hgnc_symbol
#stats <- stats[stats$symbol != "",]
stats <- merge(stats,
               conversion,
               by = 'hgnc_symbol')
ranks <- stats$`voom_all_HGNC$logFC`
stats$entrez_gene_id <- as.character(stats$entrez_gene_id)
names(ranks) <- stats$entrez_gene_id
ranks <- sort(ranks)

pathways <- reactomePathways(stats$entrez_gene_id)
test <- fgsea(pathways,exampleRanks)

data("examplePathways")
data("exampleRanks")
fgseaRes <- fgsea(pathways = pathways,stats = ranks, nperm = 10000, maxSize = 500)


pdf("../GSEA/GSC_vs_HNPC_gsea.pdf", 14,8)
topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
print(
  
  plotGseaTable(pathways[topPathways], ranks, fgseaRes, 
                gseaParam = 0.5)
)
dev.off()





```



```{r RSEM}
#read in the RSEM counts
read.counts <- read.table("../RSEM_out/rsem_isoforms_matrix.txt")


#change column names to the samples names form the data
colnames(read.counts) <- c("387++","456++","3359++","3691++","3832++","4121++","4302++","H2S++","387-","456-", "3359-","3691-","3832-","4121-","4302-","H2S-", "1123++", "528++", "1123-", "528-", "hNP1_old","hNP1_new","17231","16157")
read.counts <- subset(read.counts,
                      select = -c(`1123++`, `528++`, 
                                  `1123-`,`528-`, 
                                  hNP1_old, hNP1_new, 
                                  `17231`, `16157` ))
#organize to that the non-GSCs are the first and the control samples. GSCs are the treatments
read.counts <- read.counts[,c(9:16, 1:8)]
#factor of the cell types
celltype <- factor(c(rep("D",8),   #D stands for differentiated cell or non-GSCs
                     rep("S",8)),   #S stands for stem cell of GSC
                   levels = c("D","S"))

cellline <- factor(c(1:8, 1:8))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above
design <- model.matrix(~ cellline + celltype)   

# remove low count reads
# convert to DGE list
# normalize the data
dge_counts <- 
  read.counts[rowSums(cpm(read.counts) >= 1) >= 8,] %>%
  DGEList() %>%
  calcNormFactors(method = c("TMM"))



#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(counts = dge_counts, 
                   design = design,
                   plot=TRUE)
# MDS plot of the data
plotMDS(x = dge_counts,
        cex = 1, 
        main = "MDS of Stem vs Diff", 
        labels = colnames(read.counts), 
        top = 500, 
        dim.plot = c(1,2),
        gene.selection = 'common')


#linear fit and eBayes funtions for differential expression
fit <-
  eBayes(
    lmFit(object = voom_counts,
        design = design))

# voom fold changes and p values for all genes
voom_all <- topTable(fit = fit, 
                     coef = "celltypeS",
                     sort.by="p", 
                     number =nrow(voom_counts))
# voom fold changes and p values based on significant hits
voom_top <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 0,
                     p.value = 0.05)


# name conversion to HGNC symbols
# TODO clean up name conversions
voom_top$transcript_id <- rownames(voom_top)
voom_top_HGNC <- merge(x = voom_top, 
                       y = conversion[,c(2,3)], 
                       by = "transcript_id")

voom_all$transcript_id <- rownames(voom_all)
voom_all_HGNC <- merge(x = voom_all,
                       y = conversion[,c(2,3)], 
                       by = "transcript_id")

# write the differently expressed transcripts to file
write.table(x = voom_all_HGNC,
            file = "../data/GSC_vs_nonGSC_all_differentially_expressed_transcripts.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)
write.table(x = voom_top_HGNC, 
            file = "../data/GSC_vs_nonGSC_top_pval05_differentially_expressed_transcripts.csv",
            quote = FALSE, 
            sep = ",",
            row.names = F)


```








```{r SRP040525}

#load in the star counts
read.counts <- importStarCounts(folder = "../SRP040525_counts/", 
                                column = 4)
# remove unnecessary name data
colnames(read.counts) <- gsub(pattern = "ReadsPerGene.out.tab", "", colnames(read.counts))

#remove reads that did not get counted due to multi mapping and unalignment
read.counts <- read.counts[!rownames(read.counts)=="N_unmapped",]
read.counts <- read.counts[!rownames(read.counts)=="N_multimapping",]
read.counts <- read.counts[!rownames(read.counts)=="N_noFeature",]
read.counts <- read.counts[!rownames(read.counts)=="N_ambiguous",]
  

#select only the relevant sample sets
read.counts <- read.counts[,c(2,5,8,11)]

#factor of the ecperiment type and cell type
kd <- factor(c(rep("WT",2),   #D stands for differentiated cell or non-GSCs
               rep("KD",2)),    #S stands for stem cell of GSC
             levels = c("WT","KD"))
celltype <- factor(c("S","D", "S","D"), 
                   levels = c("D", "S"))  #Factor for the samples used. 8 paired samples
#create the design matrix based on the factors above

#cellline <- factor(c(1:3,1:3), levels = 1:3)  #Factor for the samples used. 8 paired samples

#cell <- factor(paste(celltype, cellline, sep = "."))
#design <- model.matrix(~0+cell)
design <- model.matrix(~  kd + celltype)


# removing low count reads
# convert to DGE List
# Normalize
dge_counts <-
  read.counts[rowSums(cpm(read.counts) >= 1) >= 4,] %>% 
  DGEList() %>%
  calcNormFactors(method = c("TMM"))


#voom for creating converting RNA seq reads to microaray data for the pipeline
voom_counts <-voom(counts = dge_counts, 
                   design = design, 
                   plot = T)


#MDS plot of the data
plotMDS(x = dge_counts, 
        cex = 1, 
        main = "MDS of Stem vs Diff", 
        labels = colnames(read.counts), 
        top = 500, 
        dim.plot = c(1,2), 
        gene.selection = 'common')

#fit <- lmFit(voom_counts,design)

#contrasts <- makeContrasts(((cellKD.2+cellWT.2)/2)-(cellKD.4+cellKD.4)/2  , levels = design)
#fit <- contrasts.fit(fit, contrasts)
#fit <- eBayes(fit, trend = T)

#linear fit and eBayes funtions for differential expression
fit <-  
  eBayes(
    lmFit(object = voom_counts,
          design = design))

#voom fold changes and p values for all genes
voom_all <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by="p",
                     number =nrow(voom_counts))
#voom fold changes and p values based on significant hits
voom_top <- topTable(fit = fit,
                     coef = "celltypeS", 
                     sort.by="p",
                     number =nrow(voom_counts),
                     lfc = 2,
                     p.value = 0.05)


#name conversion to HGNC symbols
voom_all_HGNC <- EnsembltoHGNC(voom_all)
voom_top_HGNC <- EnsembltoHGNC(voom_top)


 #saving the  top tables for use in other scripts
#commented out for use only when needed

#write.table(x = voom_all_HGNC, "../data/StemVsDiff_all_DE_genes.txt", quote = FALSE, sep = ",", row.names = F)
#write.table(x = voom_top_HGNC, "../data/StemVsDiff_top_DE_genes.txt", quote = FALSE, sep = ",", row.names = F)
#write.table(x = as.data.frame(voom_counts$E), "../data/star_v5_normalized_counts.txt", quote = FALSE,  col.names = TRUE, sep = ",")





```


