---
title: "MoD Microarray  (GSL, DOX)"
author: "Gurkan Bebek"
date: "March 11, 2016"
output: html_document
---
`r date()`





```{r loadRMA, echo=FALSE , message=FALSE,warning=FALSE,cache=TRUE }

####                      ####
####  MICROARRAY ANALYSIS ####
####                      ####

library(knitr)

source("http://bioconductor.org/biocLite.R")
biocLite(c("oligo","pd.hugene.2.0.st"))
library(oligo)
library(pd.hugene.2.0.st)
library(annotate)

setwd("~/Research/MoD_PR/")

celFiles <-oligoClasses::list.celfiles("data",full.names=TRUE,listGzipped = TRUE)

samples<-read.table("samples.txt",sep="\t",header=T,stringsAsFactors = FALSE)
rownames(samples)<-sub("-","_",paste(samples$Group,samples$Sample_Name,sep="_"))

theData <- data.frame(Key=samples$filename)
affyRaw <- read.celfiles(filenames = celFiles)
rownames(theData) <- basename(celFiles)
# sampleNames(affyRaw)

mt <- match(samples$filename, sub(".CEL.gz","",sampleNames(affyRaw)))

sampleNames(affyRaw)<-(row.names(samples))[mt]
vmd <- data.frame(labelDescription = colnames(samples))
phenoData(affyRaw) <- new("AnnotatedDataFrame", data = samples[mt, ], varMetadata = vmd)

#if( grepl("Gene", class(affyRaw)[1]) || grepl("Exon", class(affyRaw[1]) )) {
affyRMA <- oligo::rma(affyRaw, target="core")
#} else {
#  affyRMA <- rma(affyRaw) # no target
#} 
#dim(exprs(affyRaw))

kable(pData(affyRMA))
  
#e <- exprs(affyRMA)
e <- exprs(affyRMA)
write.csv(round(e,3),file=paste("MoD_Microarray_RMA_for_GSE.csv",sep=""))

```


```{r features,  echo=FALSE , message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="loadRMA"}
library(knitr)
## ANNOATTION
featureData(affyRMA) <- getNetAffx(affyRMA, 'transcript')
#names(fData(affyRMA))
#with(fData(affyRMA), table(seqname, category))

#head(fData(affyRMA[20200:20210,]))

fData(affyRMA)$gocellularcomponent<-NULL
fData(affyRMA)$gomolecularfunction<-NULL
fData(affyRMA)$gobiologicalprocess<-NULL

fData(affyRMA)$swissprot   <-NULL
fData(affyRMA)$unigene   <-NULL
fData(affyRMA)$pathway   <-NULL
fData(affyRMA)$proteindomains <-NULL

# (fData(affyRMA))[10000:10010,]


x<-sapply(fData(affyRMA)$geneassignment, function(x){ (strsplit(x=x,split = " // " ,fixed = TRUE)[[1]])[1]})
fData(affyRMA)$gene1<-x
x<-sapply(fData(affyRMA)$geneassignment, function(x){ (strsplit(x=x,split = " // " ,fixed = TRUE)[[1]])[2]})
fData(affyRMA)$SYMBOL<-x

#  mapped.genes <- bitr(genes, fromType = "SYMBOL", toType = "ENTREZID", annDb = "org.Hs.eg.db")


```



### QC


#### NUSE

Normalized Unscaled Standard Errors (NUSE) can  be used for assessing quality.  In
this case, the standard error estimates obtained for each probe is used.
Data is standardized  across  arrays  so  that  the  median  standard  error  for  that
genes is 1 across all arrays.  This process accounts for differences in variability between
genes.  An array where there are elevated SE relative to the other arrays is typically of
lower  quality.   Boxplots  of  these  values,  separated  by  array  can  be  used  to  compare
arrays. 

A model that accounts for probe (feature) and sample effects, whose estimates and standard errors can be recovered, respectively, with the coef and se methods, as shown below.


```{r QC, echo=FALSE, message=FALSE, cache=TRUE, comment=NA, fig.align='center',dependson="loadRMA"}

#a<-read.celfiles(filenames = celFiles)
library(oligo)
library(genefilter)
library("simpleaffy")
#library("affyPLM")  #for NUSE
   # Read cel files from directory
   # Create affy QA matrix
   # RLE and NUSE
#plmStruc = fitPLM(affyRaw)
#RLE(plmStruct, type="stats")
#NUSE(plmStruct, type="stats")
# ?fitProbeLevelModel
fit1 <- fitProbeLevelModel(affyRaw, background=TRUE, normalize=TRUE, target="core", method="plm", verbose=TRUE, S4=TRUE)
# summary(fit1)
#oligo::image(fit1)
oligo::NUSE(fit1,las=2)
oligo::RLE(fit1,las=2)
```


#### Data before and after normalization:

```{r boxes,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="loadRMA"}
#RAW

y <- round(log2(exprs(affyRaw)),3)
boxplot(y,las=2)

#AFTER RMA
boxplot(affyRMA,las=2)
boxplot((e), col=as.numeric(as.factor(pData(affyRMA)$Group))+1,las=2)

```


#### MDS: Multidimensional Scaling 
Utilize all probe measurements to plot difference between the samples. Similar samples (replicates) should be closer.  

```{r MDS,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="loadRMA"}
library(knitr)
md <- cmdscale(dist(t(e)),2)
mycolors<-c("red","green","blue","purple")[as.factor(substr(row.names(md),1,2))]
 
plot(md,xlim=c(min(md[,1])*1.5,max(md[,1])*1.3), col = mycolors)#,pch=substr(row.names(md),1,8) , las=2 ,

#install.packages("calibrate")
library(calibrate)
#textxy(md[,1], md[,2],  samples$Sample_IDs)#theData$Key)#substr(row.names(md),1,8) )
textxy(md[,1], md[,2],  lab=row.names(md))#theData$Key)#substr(row.names(md),1,8) )
```

#### FKBP5 expression with PCR
```{r QC_FKBP5, echo=FALSE, message=FALSE, cache=TRUE, comment=NA, fig.align='center',dependson="features"}

FKBP5_PCR<-matrix(ncol = 12,nrow=1,data=c( 0.02057920753, 0.02021594346, 0.01878667918, 0.03110819981, 0.03638299289, 0.0276231467, 0.05425903337, 0.06635940932, 0.07431935259, 0.06482235266, 0.07786844036, 0.0667823917))
colnames(FKBP5_PCR)<-c("6-1", "6-2", "6-3", "7-1", "7-2", "7-3", "8-1", "8-2", "8-3", "10-1", "10-2", "10-3")

kable(FKBP5_PCR,caption = "FKBP5 expression (PCR)")
FKBP5_MicroArray<-e[(fData(affyRMA)$SYMBOL %in% "FKBP5"),c(1,11,5,12,4,3,7,10,6,2,9,8)]

kable(as.data.frame(t(FKBP5_MicroArray)),caption = "FKBP5 expression (microarray)")

plot(x=FKBP5_PCR,y=t(FKBP5_MicroArray),col=substr(colnames(FKBP5_PCR),1,1),pch= 15 )
library(calibrate)
#textxy(md[,1], md[,2],  samples$Sample_IDs)#theData$Key)#substr(row.names(md),1,8) )
textxy(FKBP5_PCR,t(FKBP5_MicroArray),  lab=colnames(FKBP5_PCR)) #theData$Key)#substr(row.names(md),1,8) )

#cor(t(rbind(FKBP5_PCR,t(FKBP5_MicroArray))))
```



#### PGR expression with PCR
```{r QC_PGR, echo=FALSE, message=FALSE, cache=TRUE, comment=NA, fig.align='center',dependson="features"}

PGR_PCR<- read.delim("real_time_pcr_PGR.txt",sep="\t",header = TRUE,row.names=1) 
 

kable(PGR_PCR,caption = "PGR expression (PCR)")
PGR_MicroArray<-e[(fData(affyRMA)$SYMBOL %in% "PGR"),c(1,11,5,12,4,3,7,10,6,2,9,8)]

kable(as.data.frame(t(PGR_MicroArray)),caption = "PGR expression (microarray)")

plot(x=PGR_PCR$Rel.Exp,y=t(PGR_MicroArray),col=substr(row.names(PGR_PCR),1,1),pch= 15 )
library(calibrate)
#textxy(md[,1], md[,2],  samples$Sample_IDs)#theData$Key)#substr(row.names(md),1,8) )
textxy(PGR_PCR$Rel.Exp,(PGR_MicroArray),  lab=row.names(PGR_PCR)) #theData$Key)#substr(row.names(md),1,8) )

#cor(t(rbind(PGR_PCR,t(FKBP5_MicroArray))))
```


##### Conclusions

Based on these results, there is no sample mix-up, but the DOX is not changing the mRNA expression profiles significantly in these cell lines. This might affect top diff. expresed genes.
 

Based on QC plots, array is fine. 



### Filtering

```{r limma,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="features"}




#Will remove 6-1 and 7-2 to make sure.

# mt<-match(c("6-1","7-2"),pData(affyRMA)$Sample_Name )
# affyRMA2<-affyRMA[,-mt]
 
affyRMA2<-affyRMA 
dim(affyRMA2)

# ### 
# md <- cmdscale(dist(t(round(exprs(affyRMA2),3))),2)
# #mycolors<-c("red","green","blue","purple")[as.factor(substr(row.names(md),1,2))]
#  
# plot(md,xlim=c(min(md[,1])*1.5,max(md[,1])*1.3))#,pch=substr(row.names(md),1,8) , las=2 ,
# 
# #install.packages("calibrate")
# library(calibrate)
# #textxy(md[,1], md[,2],  samples$Sample_IDs)#theData$Key)#substr(row.names(md),1,8) )
# textxy(md[,1], md[,2],  lab=row.names(md))#theData$Key)#substr(row.names(md),1,8) )
# 
# 
###




samples <- pData(affyRMA2)
library(limma)
design <- model.matrix(~factor( samples$Group)+0)
colnames(design)<- substring(colnames(design),22,30)
contrast.matrix <- makeContrasts("Dox-Control", "GSL-Control", "Bdom-Control", levels=design)

colnames(design)<- substring(colnames(design),22,30)
fit <- lmFit(affyRMA2, design)
fit2 <- contrasts.fit(fit, contrast.matrix)
ebayes <- eBayes(fit2)
 
a1 <- topTable(ebayes, adjust="fdr",coef=1)
a2 <- topTable(ebayes, adjust="fdr",coef=2)
a3 <- topTable(ebayes, adjust="fdr",coef=3)

b1 <- topTable(ebayes, adjust="fdr",coef=1,p.value = 0.05,number = 10000 )
b2 <- topTable(ebayes, adjust="fdr",coef=2,p.value = 0.05,number = 10000 )
b3 <- topTable(ebayes, adjust="fdr",coef=3,p.value = 0.05,number = 10000 )

c1 <- topTable(ebayes, adjust="fdr",coef=1,p.value = 1,number = 100000 )
c2 <- topTable(ebayes, adjust="fdr",coef=2,p.value = 1,number = 100000 )
c3 <- topTable(ebayes, adjust="fdr",coef=3,p.value = 1,number = 100000 )






kable(a1[,-c(8,9)],caption = "Top 10 genes Dox-Control")

kable(a2[,-c(8,9)],caption = "Top 10 genes GSL-Control")

kable(a3[,-c(8,9)],caption = "Top 10 genes Bdom-Control")

```


I have analyzed data using limma (Linear Models and Empirical Bayes Methods) and then I got whether the related t-statistics are up, down or not significant for three contrasts of interest:


FDR-corrected p-values<0.05:

```{r venn,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="limma"}

results <- decideTests(fit2, method = "separate", adjust.method = "fdr", p.value = 0.05) 

## First diagram 
vennDiagram(results,include=c("up","down"),counts.col=c("red","green"), 
            main="Diff. Expressed Transcripts (adj-p-val<0.05)")

```


## Reactome/KEGG/Gene Ontology Enrichment

In these anaylses, we look for enrichment in curated gene sets.
If there is no enrichmnet for a given set, it will not be shown.

### __ _Dox, GSL and Bdom enrichment:_ __

```{r Reactome,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="limma",fig.width=11}

library(ReactomePA)
library(clusterProfiler)

mapped.genes.b1 <- bitr(b1$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")
mapped.genes.b2 <- bitr(b2$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")
mapped.genes.b3 <- bitr(b3$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")

geneLists<- list(Dox=mapped.genes.b1$ENTREZID, 
                 GSL=mapped.genes.b2$ENTREZID,
                 Bdom=mapped.genes.b3$ENTREZID)

res <- compareCluster(geneLists, fun="enrichGO",ont = "MF",pAdjustMethod = "fdr")
plot(res,title="Gene Ontology Enrichment  (Molecular Function -- FDR-adj-p-val<0.05)")
#kable(summary(res))

res <- compareCluster(geneLists, fun="enrichGO",ont = "BP",pAdjustMethod = "fdr")
plot(res,title="Gene Ontology Enrichment  (Biological Process -- FDR-adj-p-val<0.05)")


res <- compareCluster(geneLists, fun="enrichGO",ont = "CC",pAdjustMethod = "fdr")
plot(res,title="Gene Ontology Enrichment  (Cellular Component -- FDR-adj-p-val<0.05)")

res <- compareCluster(geneLists, fun="enrichKEGG",pAdjustMethod = "fdr")
plot(res,title="KEGG Pathway Enrichment (FDR-adj-p-val<0.05)")

res <- compareCluster(geneLists, fun="enrichPathway",pAdjustMethod = "fdr")
plot(res,title="Reactome Pathways Enrichment (FDR-adj-p-val<0.05)")

## formula interface
# mydf<-data.frame(Entrez=c(mapped.genes.b1$ENTREZID,mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
#                    group = c(rep('Dox',length(mapped.genes.b1$ENTREZID)),
#                              rep('GCL',length(mapped.genes.b2$ENTREZID)),
#                              rep('Bdom',length(mapped.genes.b3$ENTREZID)))
#                  )
# xx.formula <- compareCluster(Entrez~group, data=mydf, fun="enrichGO" )
# summary(xx.formula)

```


### Shared and unique genes in GSL and/or Bdom:

```{r Reactome2,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="limma",fig.width=11}

require(ReactomePA)
require(clusterProfiler)

mapped.genes.b1 <- bitr(b1$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")
mapped.genes.b2 <- bitr(b2$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")
mapped.genes.b3 <- bitr(b3$SYMBOL, fromType = "SYMBOL", toType = "ENTREZID", annoDb = "org.Hs.eg.db")

geneLists<- list(GSL_only = setdiff(mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
                 'Bdom+GSL'= intersect(mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
                 Bdom_only = setdiff(mapped.genes.b3$ENTREZID,mapped.genes.b2$ENTREZID))

res <- compareCluster(geneLists, fun="enrichGO",ont = "MF",pAdjustMethod = "fdr")
plot(res,title="Gene Ontology Enrichment  (Molecular Function -- FDR-adj-p-val<0.05)")
#kable(summary(res))

res <- compareCluster(geneLists, fun="enrichGO",ont = "BP",pAdjustMethod = "fdr")
plot(res,title="Gene Ontology Enrichment  (Biological Process -- FDR-adj-p-val<0.05)")


#res <- compareCluster(geneLists, fun="enrichGO",ont = "CC",pAdjustMethod = "fdr")
#plot(res,title="Gene Ontology Enrichment  (Cellular Component -- FDR-adj-p-val<0.05)")

res <- compareCluster(geneLists, fun="enrichKEGG",pAdjustMethod = "fdr")
plot(res,title="KEGG Pathway Enrichment (FDR-adj-p-val<0.05)")

res <- compareCluster(geneLists, fun="enrichPathway",pAdjustMethod = "fdr")
plot(res,title="Reactome Pathways Enrichment (FDR-adj-p-val<0.05)")



## formula interface
# mydf<-data.frame(Entrez=c(mapped.genes.b1$ENTREZID,mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
#                    group = c(rep('Dox',length(mapped.genes.b1$ENTREZID)),
#                              rep('GCL',length(mapped.genes.b2$ENTREZID)),
#                              rep('Bdom',length(mapped.genes.b3$ENTREZID)))
#                  )
# xx.formula <- compareCluster(Entrez~group, data=mydf, fun="enrichGO" )
# summary(xx.formula)

```

#
#
# .
# .
# .
# .
# .
#
# 

# SUPPLEMENTARY
```{r SUPPLEMENTARY,  echo=FALSE, message=FALSE,warning=FALSE,cache=TRUE,include=TRUE,dependson="limma"}


kable(mapped.genes.b1, caption = "Dox-Control Genes used in Enrichment Analysis")
kable(mapped.genes.b2,caption = "GSL-Control Genes used in Enrichment Analysis")
kable(mapped.genes.b3,caption = "Bdom-Control Genes used in Enrichment Analysis")

kable(b1[,-c(8,9)],caption = "Top diff. expressed genes Dox-Control")
write.csv(b1[,-c(8,9)],file="Top diff. expressed genes Dox-Control.csv")
write.csv(c1[,-c(8,9)],file="All genes Dox-Control.csv")

kable(b2[,-c(8,9)],caption = "Top diff. expressed genes GSL-Control")
write.csv(b2[,-c(8,9)],file="Top diff. expressed genes GSL-Control.csv")
write.csv(c2[,-c(8,9)],file="All genes GSL-Control.csv")

kable(b3[,-c(8,9)],caption = "Top diff. expressed genes Bdom-Control")
write.csv(b3[,-c(8,9)],file="Top diff. expressed genes Bdom-Control.csv")
write.csv(c3[,-c(8,9)],file="All genes Bdom-Control.csv")



# 
# geneLists<- list(GSL_only = setdiff(mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
#                  'Bdom+GSL'= intersect(mapped.genes.b2$ENTREZID,mapped.genes.b3$ENTREZID),
#                  Bdom_only = setdiff(mapped.genes.b3$ENTREZID,mapped.genes.b2$ENTREZID))
# 
# 
# kable(b1[,-c(8,9)],caption = "Top diff. expressed genes Dox-Control")
# 
# kable(b2[,-c(8,9)],caption = "Top diff. expressed genes GSL-Control")
# 
# kable(b3[,-c(8,9)],caption = "Top diff. expressed genes Bdom-Control")



```

